\documentclass[12pt]{article}

%\usepackage{a4wide}
\usepackage{amsfonts}
%\usepackage{amssymb}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{macros}

\newcommand{\mach}{$\langle$\textsl{my\_machine}$\rangle$}


\begin{document}

\title{The Gyrokinetic Plasma Turbulence Code {\sc Gene}: User Manual}
\vspace{1cm}
\date{\Large August 23, 2013}
\author{\gene Development Team}

\maketitle

\vspace{2cm}
\begin{center}
\includegraphics[width=\textwidth]{gene_logo2.png}
\end{center}

\thispagestyle{empty}
\newpage

\tableofcontents

\newpage
\section{Introduction}
\label{sec:intro}

\subsection{A few words on the code}
The gyrokinetic plasma turbulence code {\sc Gene} (this acronym stands for Gyrokinetic Electromagnetic
Numerical Experiment) is a software package dedicated to solving the nonlinear gyrokinetic
equations in a flux-tube domain. Alternatively, it can be operated in a linear mode, thus
calculating the properties (like complex frequencies, parallel mode structures, and quasilinear
transport coefficients) of the microinstabilities driving the turbulence. Furthermore, a solver for the neoclassical
equilibrium is implemented. {\sc Gene} can be run on
a large number of different computer architectures, including Linux clusters and various massively
parallel systems, using anything between a single and tens of thousands of processors. While the code
itself is written in {\tt Fortran90/95/03}, the package also includes an {\tt IDL} based tool for data
visualization and analysis and a python user interface for setting up simulations. {\sc Gene} has been developed 
by a team of people (the Gene Development Team, led by F.~Jenko) over the last several years, and is 
freely available to anyone accepting the general guidelines stated in the ``User Agreement'' document.


\subsection{A few words on the basic equations}

In the magnetic confinement fusion (MCF) community, it is widely accepted that {\em ab initio}
simulations of turbulence in the core of tokamaks and stellarators are to be based on the
nonlinear gyrokinetic equations as first derived in the 1980s by various authors \cite{frieman82,
hahm88,hahmbriz88,brizard89} (see also the recent review, Ref.~\cite{GK07}). As it turns out, the fast
gyromotion can be removed from the basic equations analytically, such that each particle species
is described by a time-dependent distribution function in a five-dimensional phase space:
\begin{itemize}
\item the electron distribution function, $f_e({\bf R},v_\|,\mu,t)$,
\item the (main) ion distribution function, $f_i({\bf R},v_\|,\mu,t)$,
\item the distribution functions of other -- active or passive -- particle species.
\end{itemize}
Here, ${\bf R}$ describes the position of the gyrocenter in real space, and for the velocity
space coordinates, we have chosen the parallel velocity $v_\|$ and the magnetic moment $\mu$
(other choices would also be possible, slightly affecting the form of the basic equations).

In addition, one has to advance two purely spatial, scalar quantities characterizing the
electromagnetic fields by solving modified versions of Maxwell's equations. These quantities are:

\begin{itemize}
\item the electrostatic potential, $\phi({\bf x},t)$,
\item the parallel component of the vector potential, $A_\|({\bf x},t)$, which is linked to the
  perpendicular magnetic field perturbations.
\item the parallel component of the magnetic field perturbations, $B_\|({\bf x},t)$.
\end{itemize}

The nonlinear gyrokinetic equations as used in the {\sc Gene} code
may be found, e.g., in Refs.~\cite{gene1,gene2} and in the PhD theses being available at 
Ref.~\cite{GENE-webpage}. The code currently contains:
\begin{itemize}
\item an arbitrary number of fully gyrokinetic particle species, active or passive
\item magnetic field fluctuations perpendicular and parallel to the background field
\item collision operators for each particle species, including energy scattering
\item general (tokamak or stellarator) MHD equilibria, a consistent circular model, the 
  standard $\hat s-\alpha$ geometry, and Miller local equilibria
\end{itemize}
This list is not exhaustive, however, and the code is still being extended and generalized.
From time to time, new {\sc Gene} releases will therefore be provided.

\paragraph{Neoclassical equilibrium:}

Often, turbulent processes dominate the transport of energy, particles, and momentum
in the core of fusion devices. However,in some cases, neoclassical effects may play a role.
In the flux-tube limit, the neoclassical equation is contained in the
gyrokinetic equation
as its perturbed, but non-fluctuating part and it decouples from turbulence.
The neoclassical equilibrium can thus be solved separately, which is
featured by {\sc Gene}, whereby no electric/magnetic field perturbations or background
electric fields are considered.

\subsection{A few words on magnetic geometry}

As is explained in Ref.~\cite{beer}, one can save 2-3 orders of
magnitude in computational effort -- compared to more traditional
ways of treating the geometry -- by using a coordinate system
which is aligned to the background magnetic field. This is because
the turbulent fluctuations have very large correlation lengths
along the field lines (here, the particle motion is practically
free) but very small ones in the perpendicular directions (due to
the slowness of the gyrocenter drifts). Consequently, the parallel
structure of the fluctuations can be represented on a very coarse
grid in the field-line following direction, an advantage which
does not exist in other, non-field-aligned coordinate systems.
Concerning the simulation volume, it turns out to be most natural
and efficient to employ a flux-tube domain which follows the field
lines and whose perpendicular widths exceed the respective
correlation lengths \cite{beer}. This approach helps to minimize
the computational cost and allows for arbitrarily long simulation
times (there is no profile relaxation, provided one uses periodic
radial boundary conditions).

As a result of the use of field-aligned, non-cartesian
coordinates, several differential and integral operators occurring
in the gyrokinetic equations have to be expressed in terms of the
metric coefficients \cite{tracer}. Here, one can currently choose
between two options when running {\sc Gene}. If one is interested
in more basic tokamak physics issues, it might be sufficient to
employ a simple $\hat s$-$\alpha$ MHD equilibrium for which all
the relevant metric quantities can be calculated analytically. It
describes a large aspect ratio tokamak with circular cross
section, but allows for finite Shafranov shifts. One should be aware however
that this model contains some order $\epsilon=r/R$ inconsistencies; for 
simulations with large values of $\epsilon$ (typical greater than 0.15),
a corrected circular concentric flux surface model is available. Shaping
studies can be carried out by using the local Miller equilibrium model. If one 
aims at even more realism (e.g., shaped equilibria) or at doing stellarator
work, however, one should rather provide an appropriate input file
for {\sc Gene} containing the required geometrical information.
This can be done by means of software tools which are also
provided within the {\sc Gene} package, post-processing MHD
equilibrium data. 
Another option is to provide an EFIT g-eqdsk file, from which the 
code will automatically extract the appropriate geometric information.


\subsection{A few words on algorithmic issues}

For the solution of the nonlinear gyrokinetic equations, the {\sc Gene} code employs a fixed grid
in phase space. It may therefore be called a ``continuum'' or ``Vlasov'' or ``Eulerian'' code --
in contrast to particle-in-cell (PIC) or semi-Lagrangian codes which are based on computing the
characteristic curves in phase space. Some of the key features of the Eulerian approach are the
following:
\begin{itemize}
\item it allows for arbitrarily long simulation times
\item it exhibits very good convergence properties
\item it is very flexible with respect to algorithms
\item it can lead to highly parallelizable codes
\end{itemize}

The discretization scheme used in {\sc Gene} is based on the so-called ``method of lines.'' This
means that the phase space (differential and integral) operators are discretized first, and that
an appropriate time stepping scheme is chosen independently in a second step. The differential
operators involving the field-line following coordinate $z$ or the parallel velocity $v_\|$ are
discretized with the help of a fourth-order Arakawa scheme \cite{Arakawa66}, while the
perpendicular $x$ and $y$ directions are treated in the framework of a pseudospectral approach.
The latter implies that all linear terms (as well as spatial derivatives) are evaluated in
$k_x$-$k_y$ space, while the nonlinearities are computed in real space for better efficiency.
In this context, one has to employ Fourier transformations, of course, as well as an appropriate
dealiasing method to avoid spurious couplings of the smallest and largest scales. The velocity
space integrations are done via Gauss and trapezoidal rules in $\mu$ and $v_\|$ space, respectively.
Collisions are modeled in GENE using a linearized Landau-Boltzmann operator, which is discretized
by means of a finite volume scheme. Most of the above numerical methods are explained in standard 
text books on applied mathematics or computational fluid dynamics (CFD).

The time stepping, on the other hand, is done by means of a fourth-order explicit Runge-Kutta method
which operates very close to the theoretical optimum (in terms of minimum computational cost per
time step unit) of any explicit scheme for our given problem. Unless the code user decides to pick
another value, the initial time step is chosen automatically by {\sc Gene} such that it marginally
satisfies the linear stability limit -- often set by the parallel electron dynamics in multi-species
runs. This is done by an eigenvalue solver which is invoked as part of the initialization process.
However, even when starting a simulation at sufficiently low fluctuation amplitudes, the latter are
likely to get big enough in the course of the simulation such that the time step needs to be adapted
(by the code itself) in order to account for Courant-L\'evy-Friedrichs type time step restrictions
set by the nonlinear terms. Such nonlinear limits would even have to be respected in a purely implicit
code (which does not exist), since they are also to be regarded as a crucial indicator of {\em accuracy}
in the context of turbulent plasma dynamics. Thus, when doing nonlinear runs with {\sc Gene}, one is
usually operating close to the point of maximum efficiency.
Starting with GENE release 1.6, the time stepping of the collisional dynamics is separated from the 
rest of the Vlasov dynamics by means of an adaptive operator scheme, usually reducing the number of
evaluations per timestep for the collision operator.

Alternative to time stepping, which solves the gyrokinetic equation as an initial value problem,
the linearized equation allows for eigenvalue computations. To that aim, interfaces 
to the PETSc/SLEPc package and several other linear algebra packages are implemented.
The neoclassical equilibrium can be obtained by initial value computations or by an algebraic solver.
{\sc Gene} features an interface to the PETSc package for that purpose.

\subsection{A few words on computer resources}\label{sec5}

{\sc Gene} has been parallelized and ported to a large number of different computer architectures.
The latter include various Linux clusters, Mac OS, as well as massively parallel systems of the following
type:
\begin{itemize}
\item IBM Power 4-6
\item IBM BlueGene/L and P
\item Cray XT3-6
\item various Intel architectures
\end{itemize}
It must be stressed here, however, that this list is neither exhaustive nor final. Thus, if any
(potential) user of the code would like to know if {\sc Gene} also runs on his or her favorite
architecture not mentioned above, the {\sc Gene} Development Team should be contacted
(\href{mailto:gene@ipp.mpg.de}{gene@ipp.mpg.de}).

While {\sc Gene} is highly parallelizable and has already been run on tens of thousands of
processors, there are many interesting applications which do not require such enormous resources.
For example, linear runs -- determining properties of the microinstabilities driving the turbulence
(like complex frequencies or quasilinear transport coefficients) -- can be done even on a single
processor and do not take much time. Moreover, for certain studies, it might be possible to slightly
change some physical or numerical parameters (like the plasma beta, the ion-to-electron mass ratio,
the temperature ratio, the simulation box size, or the phase space resolution) such that the
computational effort can be reduced significantly without affecting the results too much. This
clearly needs to checked, however, on a case-by-case basis (one such example is mentioned in
Ref.~\cite{gene2}). If it indeed turns out to be feasible to do 'reduced simulations' of this sort
in any given context, the main load of a project may thus be carried, e.g., by a Linux cluster
or another small-size parallel platform. Otherwise, one needs to resort to more powerful machines.
Finally, there are other applications -- like nonlinear multi-scale or stellarator simulations --
which {\em always} require a very significant amount of computer resources. Projects involving such
demanding runs can only be initiated if access to adequate high performance computers is provided.





\newpage

\section{Installing Gene on your machine}

\subsection{Getting the source code}

{\sc Gene} is stored in a password protected {\tt svn} repository that is accessible via the Internet
(\url{https://solps-mdsplus.aug.ipp.mpg.de/wsvn}). If you do not have an account and password already,
please contact the \gene Development Team (\href{mailto:gene@ipp.mpg.de}{gene@ipp.mpg.de}).
To download {\sc Gene} to your local system, make sure that you have {\tt svn} installed, go to the
directory in which you want to install the code (e.g.~\texttt{/gene}) and type
\begin{verbatim}
 svn checkout https://solps-mdsplus.aug.ipp.mpg.de/repos/
 GENE11/tags/release-<current release number>
\end{verbatim}
As a result, you should get the release files containing:
\begin{itemize}
\item the \gene source code (located in the \texttt{./src} directory)
\item the {\tt IDL}-based postprocessing routines (in \texttt{./diagnostics})
\item the Python-based launcher tool (in your \texttt{/gene} root directory) 
\item the \gene documentation (in \texttt{./doc})
\item the \gene test suite (in \texttt{./testsuite})
\end{itemize}
as well as the {\tt newprob} script, several tools for managing {\sc Gene} runs (in \texttt{./tools}), 
a generic {\tt makefile}, and several machine-specific makefiles (in \texttt{./makefiles}).
To receive updates later on, simply go to your \gene base directory (here: \texttt{/gene}) and type
\verb|gmake update VERSION=Y.Z| where Y.Z is the new release version number. \\
More information about {\tt svn} commands can be found at \\
\url{http://svnbook.red-bean.com}.

\subsection{Creating a new machine-specific makefile}

Machine-specific makefiles and batch scripts for the massively parallel systems mentioned in Sec.~\ref{sec5} and for a
number of Linux clusters (at RZG Garching, at EPFL Lausanne, and at JET) are provided automatically. If
you are planning to use \gene on a different architecture, you will have to create a new machine-specific
{\tt makefile}. For this purpose, the template provided in \texttt{./makefiles/new\_machine} should
be self-explanatory. Note that the makefile structure has been changed with release 1.6 to provide common 
definitions/flags for various compilers (to be found in \texttt{./makefiles/compilers}), and also a separate
file containing general compiling rules (\texttt{./makefiles/rules.mk}). 
The external software you definitely need to specify in order to compile \gene are:
\begin{itemize}
\item a {\tt Fortran2003} conform compiler 
\item the {\tt MPI}2 message passing interface
\item an FFT routine (FFTW, MKL, or ESSL)
\item the BLAS/LAPACK library (also contained in MKL and ESSL)
\end{itemize}
Optional software packages that can extend the functionality of \gene are 
\begin{itemize}
%\item {\tt OpenMP}
\item for eigenvalue computations and a precise determination of the maximum linear time step,
the {\bf \emph{complex}} version of the (free) PETSC/SLEPC package 
\cite{petsc1,petsc2,slepc}, available at \url{http://www.mcs.anl.gov/petsc} 
and \url{www.grycap.upv.es/slepc} {\bf (highly recommended!)}
\footnote{set {\tt --with-scalar-type=complex} at configuration and furthermore {\tt --with-precision=double} if \gene is compiled with double precision}
\item for full eigenvalue spectra, the ScaLAPACK library, available at \url{http://www.netlib.org/scalapack/}
\item for the HDF5 output format the corresponding library, see \url{http://www.hdfgroup.org/HDF5/}
\end{itemize}
In order for the {\tt makefile} to work properly, choose a name for
your machine (\mach\ in the following) and rename the directory and {\tt makefile} from\\
\texttt{./makefiles/new\_machine/new\_machine.mk}\\ to\\ \texttt{./makefiles/\mach/\mach.mk}\\
You can also save submit script templates for your machine in the \mach/ folder; they will then be
available in every problem folder you create.

NOTE: The machines indicated above are automatically recognized by the values of certain predefined
environment variables. If you have written a new machine-specific makefile, you either have to
extend these routines (located in {\tt ./makefiles/machine.def})
or to set an environment variable MACHINE to \mach\ before you can compile or test your installation.

\subsection{Testing your installation}
Once you have a (preliminary) makefile for your machine, go to the \texttt{./testsuite} directory
and check whether you can compile with\\
\qquad\verb|gmake -f ../makefile| \\
Note that the first call to gmake will generate a directory named \texttt{./bin}, which contains
the machine-specific makefile and, after compilation, an executable named \texttt{gene\_\mach}.
This local version of the makefile can then be improved and tested before it is finally transferred back to
\texttt{./makefiles/\mach}.
After you manage to compile without errors, type \\
\qquad\verb|./testsuite| \\
This will compute test problems with increasing complexity and
compare the results and time needed for the computation with reference values.
Per default, the maximum number of MPI processes used is 8, this can be decreased by the option
-mpiprocs \# MPI processes. The OMP parallelization (if supported by your machine and compiler) can
be tested by simply by setting the OMP\_NUM\_THREADS environment variable to a value greater 1 (you
may want to limit the number of MPI processes in that case, since the total number of processes used is
OMP\_NUM\_THREADS times mpiprocs).
Since release 1.5, it is possible to compile also from the \textsc{Gene} root directory
by typing \verb|gmake|. When compiling from within a \texttt{prob} directory, the position of the main
makefile has to be given by typing \verb|gmake -f ../makefile|, as before. With this new structure,
only one copy of the \textsc{Gene} executable will reside in \texttt{./bin}, instead of
separate ones for each problem directory. \\
Note, that changes to the machine specific makefiles in \texttt{./makefiles/\mach}
(e.g., after an update) need to be transferred by hand (either by simply calling \texttt{gmake distclean},
by copying the \texttt{\mach .mk} file or removing the \texttt{./bin} directory) in order to allow 
the user to keep care of local modifications.

\subsection{Using the CHEASE interface module}
The metric and equilibrium quantities used in \textsc{Gene} can be obtained using an
interface with the MHD equilibrium code CHEASE. This interface requires the
library FUTILS, which will automatically be downloaded from an svn repository, as well
as the HDF5 library. More information on these two libraries can be found  in 
\texttt{external/futils-gene/docs/futils.pdf}. To run \textsc{Gene} with this interface
the precompilation option \texttt{FUTILS=yes} should be set in the
machine.mk file.

\subsection{Installing and running \textsc{Gene} on Mac OS}

Note: this method of installing and running \gene on a Macintosh has
been tested on an Intel Mac running OS 10.4.11.

The following software has to be downloaded and installed separately
(if it has not been already): Apple XCode (for the BLAS routines), a
Fortran compiler and MPI (e.g., gfortran and MPICH, see
\texttt{http://hpc.sourceforge.net/}), and an FFT (e.g., FFTW, see
\texttt{http://www.fftw.org/install/mac.html}). For post-processing, use
can be made of the IDL Virtual machine which is available for Mac. If
uncomplicated access to SVN is desirable, an installation package can
be found at \texttt{http://homepage.mac.com/martinott/}.

Once these requirements are met, the actual installation of \gene does
not differ much from that on any other machine. Make sure that MPI is
included in the \texttt{PATH}. When accessing the makefile (i.e., when
running \texttt{newprob} or when compiling), export the OS type
beforehand: \texttt{export OSTYPE}. And before running \gene, start
the parallel environment via \texttt{mpdboot --ncpus=[CPU \#]}.

\newpage


\section{Setting up a computation}

To create a problem folder (\texttt{./prob01/}) in which the parameters for a computation are defined,
execute \\
\qquad\verb|./newprob| \\
in the \gene base directory. Successive calls to \verb|newprob| will generate new problem folders:
(\texttt{./prob02/}, \texttt{./prob03/} etc.). The problem folders may be renamed, the directory
structure must not be changed, however.

The next step is to adapt the switches for libraries, precision, optimization etc. in the machine-specific
makefile (if necessary). These switches have to be set before compilation -- later changes are only effective
after a recompilation of the code. Compile \gene by typing \\
\qquad\verb|gmake -f ../makefile| \\
in the problem directory (or by typing \verb|gmake| in the \texttt{./} directory. This action should create 
an executable called ./bin/\texttt{gene\_\mach}, which can then
be used for both linear and nonlinear runs. Unless one wants to change or update the source files, there is
usually no need to recompile the code.

The other input file to be adapted is the \texttt{parameters} file, in which various physical and numerical 
parameters can be specified. Since release 1.4, a tool named \texttt{GENE-GUI.py}, which is described 
in the next section, is available to simplify this task.

%Here, one can specify various physical
%and numerical parameters, and choose between several options concerning the input and output of data. This
%file doesn't have to be known until run time, i.e., it can still be changed after a run has been submitted
%to a queue as long as it has not yet started running. An overview of the options in the \texttt{parameters}
%file is given in section \ref{subsec:parameters}.

\subsection{The \gene launcher}
In order to provide easier user access to the \gene code, a graphical user interface is available that allows
the user to read, edit and write 'parameters' files, while providing consistency checks to prevent common 
errors. 
The interface requires a working installation of Python 2.6+ and the Python/Tkinter widget package, which
is contained in many distributions of Python (including the standard one from \url{www.python.org}). To open the program,
just call './GENE-GUI.py' from the \gene root directory. 

\subsubsection{Description of the interface}
The launcher window is divided into several tabs which sort the input parameters according to their purpose: 
\begin{description}
\item[Operation:] This tab allows the user to adjust the \gene code's operation modes and 
 should always be visited first, since changing between linear/nonlinear, initial value/eigenvalue mode or
 neoclassical solver will affect the options that can be selected in other tabs.
\item[Input/Output:] Here the user can specify output directories and the frequency with 
  which the code's output files will be written to disk.
\item[Domain:] This tab allows the user to set the grid sizes and resolutions for the simulation.
\item[General:] The 'General' tab contains parameters that affect the runtime of the 
  simulations, as well as timestep-related parameters. Furthermore, adjustments for eigenvalue 
  runs and various physics parameters can be made. 
\item[Collisions:] Collisionality and options for the collision operator can be adjusted here.
\item[Geometry:] The 'Geometry' tab contains all geometry-related parameters.
\item[Species:] In the 'Species' tab, the user can adjust the number of simulated species as 
  well as their parameters and whether they should be treated as passive tracers. In addition,
  the settings for various profile interfaces can be adjusted here.
\item[Ref. Values:] This tab allows the user to specify the reference quantities used for 
  normalization. Once entered, an automatic calculation of dimensionless parameters like 
  $\beta$, the collisionality or the debye length is possible. \gene passes 
  the reference values to the output parameters file, from where they may be used to convert 
  heat flux values to SI units in the diagnostic tool.
\item[Development:] Here the user can specify parameters which are not hard-wired in the Launcher 
  tool. For each parameter, one can choose the namelist it should be appended to.
\end{description}

In the top row of the interface, numbered buttons allow the user to manage multiple jobs in 
the same launcher instance by changing between the numbered jobs. All parameter settings that 
have been made are saved in memory and can be reloaded by returning to the same job number again. 

When hovering over labels in the user interface, tooltips will often be displayed to help the user 
decide how to adjust the respective parameter. In addition to this, the tooltip contains the 
name of the parameter as it appears in the \texttt{parameters} file, should the user wish to 
edit the latter by hand.

\paragraph{Command buttons:}
\begin{description}
\item[Read parameters:] Reads existing parameters files from the 'prob' directories as well as those 
created by \textsc{Gene}.
\item[New 'prob' dir.:] Creates a new 'prob' directory.
\item[Write parameters:] Overwrites an existing parameters file with one created by the launcher tool. 
  The user must select an existing 'prob' directory, in which the \texttt{parameters} file is to be written.
\item[Save as:] This command allows saving the parameters to an arbitrary file name.
\item[Check:] Performs some consistency checks on the input values the user has given. This is also the
case when the 'Write' button is clicked.
\item[Submit:] Submits a single run taking into account the number of processes. This function relies on 
the existence of a script named \texttt{launcher.cmd} in the \texttt{makefiles/\mach/} directory. This 
file should contain a job submission script appropriate for your machine, with certain parameters (like 
the number of MPI processes) replaced by keywords. The latter will then be replaced by the Launcher with 
the required values. See, e.g., \texttt{makefiles/bob\_cluster/launcher.cmd} for an example file, and 
the next paragraph for a list of keywords.
\end{description}

\paragraph{Keywords for job submission:}
The following keywords are machine-specific input parameters which are required to be able to submit 
jobs automatically through the Launcher. 
\begin{description}
  \item[MAXWT:] Defines the maximum wallclock time in hours.
  \item[PROCSPERNODE:] Maximum number of MPI processes per node.
  \item[SUBMITCMD:] The command required to submit a job. 
 \end{description}
These keywords have to be uppercase and defined in the form \texttt{\#\#\#~<keyword> <value>}.\\
The following keywords are intended for output and will be replaced with their appropriate values:
\begin{description}
  \item[NMPIPROCS:] Number of MPI processes
  \item[NPROCSDIV4:] Number of MPI processes, divided by four (required e.g. on BlueGenes in Virtual 
    node mode)
  \item[PPN:] Processes per node (not required by all machines)
  \item[NODES:] Number of nodes (not required by all machines)
  \item[WALLTIME:] The wallclock time limit for the simulation to be submitted (can be entered in a 
    dialog which appears after clicking the 'Submit' button).
  \item[JOBNAME:] Name of the job (can also be entered in a dialog). 
\end{description}
In case additional keywords, or a different treatment of the existing ones, are required to set up 
job submission for a given machine, the \gene development team should be notified (\texttt{gene@ipp.mpg.de}).

\paragraph{Parameter scans:}
Scans over one or more parameters can be performed by selecting the option 'Parameter scan' in the 
'Operation' frame and giving an entry like "1.0 !scan: 0.1,0.1,1.0" (the first number is necessary) 
to the parameter which should be scanned. 
This entry will be copied to the parameters file, so that all options the scanscript offers are 
available also in the launcher (see section \ref{sec:scanscript} for a description of the scanscript). 

\paragraph{Further options:}
The 'clear form' button deletes all parameters entered into the user interface, while the 'default 
values' button fills the entries so that a standard cyclone base case run (using, however, the 
corrected circular model) can be started. Note that the latter button delivers different default 
values depending on whether the nonlinear or linear mode is selected. Finally, a checkbutton 
labeled 'Advanced options' allows the user to set additional parameters that are hidden otherwise, 
while the button labeled '?' brings up the \gene documentation.

\newpage
\subsection{The \texttt{parameters} file} \label{subsec:parameters}

The \texttt{parameters} file consists of namelists containing key-value pairs.
The keywords and their meanings are given in the following list.
In parentheses, the data type ([int, real, str, bool] for integer, real, string and logical)
and default value (if any) is given. Note that strings have to be provided using quotation marks.
Parameters with default values don't have to be specified in the \texttt{parameters} file.
A newly created problem folder contains a \texttt{parameters} file for a ``Cyclone base case''
scenario described in Ref.~\cite{cyclone}.

\subsubsection{The \texttt{parallelization} namelist}
\begin{description}
\hypertarget{parallelization_nml}{} 
\item[\texttt{n\_procs\_s [int 1]:}] number of MPI processes over which the species are distributed; 
to ensure load balancing, the parameter \texttt{n\_spec} (see below) should be equal or at least divisible 
by this number; 
if set to a number smaller than one, GENE tries to find the best combination using up to as many
MPI processes as are left in the MPI\_COMM\_WORLD. The range can be restricted by additionally providing 
lower and upper integer-valued boundaries via \texttt{min\_nps} and \texttt{max\_nps}.
\item[\texttt{n\_procs\_w [int 1]:}] number of MPI processes over which the $\mu$ (magnetic moment) grid
is distributed; to ensure load balancing, the parameter \texttt{nw0} (see below) should be divisible by
this number; 
if set to a number smaller than one, GENE tries to find the best combination using up to as many
MPI processes as are left in the MPI\_COMM\_WORLD. The range can be restricted by additionally providing 
lower and upper integer-valued boundaries via \texttt{min\_npw} and \texttt{max\_npw}.
\item[\texttt{n\_procs\_v [int 1]:}] number of MPI processes over which the $v_{\parallel}$ (parallel
velocity) grid is distributed; to ensure load balancing, the parameter \texttt{nv0} (see below) should
be divisible by this number; 
if set to a number smaller than one, GENE tries to find the best combination using up to as many
MPI processes as are left in the MPI\_COMM\_WORLD. The range can be restricted by additionally providing 
lower and upper integer-valued boundaries via \texttt{min\_npv} and \texttt{max\_npv}.
\item[\texttt{n\_procs\_y [int 1]:}] number of MPI processes over which the $k_y$ ('poloidal wavenumber')
grid is distributed; to ensure load balancing, the parameter \texttt{nky0} (see below) should be divisible
by this number; if no thread-safety is provided, OMP\_NUM\_THREADS and \texttt{n\_procs\_y} cannot be chosen to be
greater than unity simultaneously; 
if set to a number smaller than one, GENE tries to find the best combination using up to as many
MPI processes as are left in the MPI\_COMM\_WORLD. The range can be restricted by additionally providing 
lower and upper integer-valued boundaries via \texttt{min\_npy} and \texttt{max\_npy}.
\item[\texttt{n\_procs\_z [int 1]:}] number of MPI processes over which the $z$ (field-line following
coordinate) grid is distributed; to ensure load balancing, the parameter \texttt{nz0} (see below) 
should be divisible by this number; 
if set to a number smaller than one, GENE tries to find the best combination using up to as many
MPI processes as are left in the MPI\_COMM\_WORLD. The range can be restricted by additionally providing 
lower and upper integer-valued boundaries via \texttt{min\_npz} and \texttt{max\_npz}.
\end{description}



\subsubsection{The \texttt{box} namelist}
\begin{description}
\item[\texttt{n\_spec [int]:}] number of particle species to be used in the computation; if only a single
ion (electron) species is employed, {\sc Gene} assumes that the electrons (ions) are adiabatic; note
that there is no upper limit to the number of particle species to be specified, and that every one
of them is treated gyrokinetically, retaining finite gyroradius effects
\item[\texttt{nx0 [int]:}] number of grid points in the $x$ direction ($x$ is the flux-surface label,
i.e., the radial coordinate); should be a power of 2; for nonlinear runs, typical values are 64 or 128,
translating into a radial resolution of about one ion gyroradius
\item[\texttt{nky0 [int]:}] number of Fourier modes in the $y$ direction ($y$ is the field-line label within
a given flux surface, i.e., the bi-normal coordinate); should be a power of 2; for nonlinear runs, typical
values are 16, 32, or 64; usually, it is desirable that the maximum value of $k_y$ (in normalized units)
is at least of the order of unity -- otherwise, the run might not be converged, resulting, e.g., in wrong
spectral decays and wrong transport levels; for linear runs, \texttt{nky0=1} together with scans 
(see Sec.~\ref{sec:scanscript}) over \texttt{kymin} are recommended 
\item[\texttt{nz0 [int]:}] number of grid points in the $z$ direction ($z$ is the field-line following
coordinate, i.e., the parallel coordinate); should be even; typical values are in the range between 16
and 32 for tokamak runs, and of the order of 100 or even more for stellarators like Wendelstein 7-X
\item[\texttt{nv0 [int]:}] number of grid points in the $v_\|$ direction ($v_\|$ is the parallel velocity);
should be even; typical values are in the range between 32 and 64, but in some cases, larger values
might be required
\item[\texttt{nw0 [int]:}] number of grid points in the $\mu$ direction ($\mu$ is the magnetic moment);
typical values are in the range between 8 and 16, but in some cases, larger values might be required
\item[\texttt{lx [real]:}] extension of the simulation box in the $x$ direction, in units of gyroradii
(calculated by using the units of mass and temperature); this value is adjusted automatically for
finite magnetic shear ($\hat s\ne 0$) such that the quantization rule
$\texttt{kymin}\cdot\texttt{shat}\cdot\texttt{lx}\in\mathbb{N}$ (see, e.g., Ref.~\cite{beer}; this rule is
necessary to make the parallel and radial boundary conditions consistent; \texttt{kymin} and \texttt{shat}
are defined below) is satisfied -- and it is neglected if the parameter \hyperlink{nexc}{\texttt{nexc}} (see below) is specified
\item[\texttt{x0 [real]:}] For the 'tracer\_efit'/'chease' geometry interfaces or when reading profile files, 
this parameter determines which 
flux surface (labeled by $\rho_{tor}$, i.e. $0<$\texttt{x0}$<1$) will be selected for simulation. When 
running with parallel flow shear and 'tracer' geometry, this parameter also has to be specified.
\item[\texttt{kymin [real]:}] minimum value of $k_y$, in units of inverse gyroradii; this corresponds to
an extension of the simulation box in the $y$ direction of $2\pi/\texttt{kymin}$, in units of gyroradii;
typically, values of the order of 0.1 or less are used for this parameter, corresponding to box sizes of
at least 60 gyroradii; inspection of the \gene diagnostics (possibly together with convergence tests) are
usually necessary to find out if a given run may be considered converged in \texttt{kymin} (if not, it
has to be lowered)
\item[\texttt{lv [real]:}] extension of the simulation box in the $v_\|$ direction, in units of the
thermal velocity (note the factor of 2 in this definition) $v_{Tj}=(2T_{j0}/m_j)^{1/2}$ of the respective
species $j$; in other words, the parallel velocity points span the range from $- \mathtt{lv} \cdot v_{Tj}$
to $\mathtt{lv} \cdot v_{Tj}$; typically,
this parameter should be at least in the range of 3 in order to make sure that the integral of the discretized,
normalized Maxwellian is sufficiently close to unity
\item[\texttt{lw [real]:}] upper end of the simulation box in the $\mu$ direction, in units of
$T_{j0}/B_\rf$ of the respective species $j$; typically, this parameter should be at least
in the range of 9 in order to make sure that the integral of the discretized, normalized Maxwellian is
sufficiently close to unity
\item[\texttt{adapt\_lx [bool t]:}] sets the optimal lx, so that the relation
  $\texttt{lx}=1/(\texttt{ky}\cdot\texttt{shat})$ holds for all ky. This maximizes the number of poloidal 
  connections for a given nx0, and means that only one linearly independent mode is considered. This parameter 
  is ignored for nonlinear runs.
\item[optional:]
\item[\hypertarget{nexc}{\texttt{nexc [int]:}}] can be used in lieu of \texttt{lx} to define the radial box width for finite
magnetic shear ($\hat s\ne 0$) via the relation $\texttt{lx}=\texttt{nexc}/(\texttt{kymin}\cdot\texttt{shat})$;
if specified, the value for \texttt{lx} is overwritten
\item[\texttt{ky0\_ind [int 1]:}] shift of the $k_y$ range for linear runs; the smallest value of $k_y$ in the computation
is then $\texttt{ky0\_ind}\cdot\texttt{kymin}$ instead of \texttt{kymin}; however, if $\texttt{nky0}>1$, neighboring
$k_y$ values are still separated by \texttt{kymin}; is neglected for nonlinear simulations
\item[\texttt{kx\_center [real 0.0]:}] shift of the $k_x$ range for linear runs, useful to investigate the 
growth rate spectrum at finite $k_x$; this parameter is neglected for nonlinear simulations
\item[\hypertarget{n0\_global}{\texttt{n0\_global [int]:}}] toroidal mode number of the \texttt{kymin} mode (is taken
  into account in the parallel boundary condition and effectively shifts the mode rational surfaces radially; typically, 
  not important for local simulations themselves but for post-processing, e.g., a proper mapping on a toroidal geometry);
  set to -1111 to automatically evaluate {\tt n0\_global = nint(kymin*C\_y/(rho\_ref/L\_ref))} next to the given {\tt kymin};
  note that \hyperlink{rhostar}{\tt rhostar} is required.
\item[\texttt{adapt\_ly [bool f]:}] adapts {\tt ly}/{\tt kymin} to be consistent with {\tt n0\_global} quantization explained
 above which might be necessary for post-processing 
 (in the local code, you are allowed to select {\tt kymin} values which do not translate to integer valued mode 
 numbers; {\tt adapt\_ly} is always enabled in global simulations)
\end{description}

\subsubsection{The \texttt{in\_out} namelist}
\label{subsubsec:inout}
\begin{description}

\item[\texttt{diagdir [str]:}] directory in which the output files should be written; it is sufficient
to provide the path relative to the directory in which the executable resides; note that quotation marks
have to be used here to identify the input as a string
\item[\texttt{chptdir [str]:}] directory in which the checkpoint file should be written; it is sufficient
to provide the path relative to the directory in which the executable resides; note that quotation marks
have to be used here to identify the input as a string
\item[\hypertarget{istep_field}{\tt istep\_field [int 0]:}] number of time steps between two consecutive 
entries into the \texttt{field.dat} output file (\hyperref[subsec:field-file]{see below}); 
this number should probably not be too small (e.g., in the range of 100-1000)
\item[\hypertarget{istep_mom}{\tt istep\_mom [int 0]:}] number of time steps between two consecutive 
entries into the \texttt{mom.dat} output file (\hyperref[subsec:mom-file]{see below}); 
this number should probably not be too small (e.g., in the
range of 100-1000); multiples of \texttt{istep\_field} are recommended
\item[\hypertarget{istep_nrg}{\tt istep\_nrg [int 10]:}] number of time steps between two consecutive 
entries into the \texttt{nrg.dat} output file (\hyperref[subsec:nrg-file]{see below}); 
this number can be rather small (e.g., in the range of 10) since this particular file will usually 
not get too big, anyways 
\item[\hypertarget{istep_omega}{\tt istep\_omega [int 20]:}] number of time 
steps after which frequency and growth rate are calculated and checked for achievement of desired precision 
(see \hyperlink{omega_prec}{\tt omega\_prec}); istep\_omega is set to 0 for nonlinear simulations or linear simulations with 
adapt\_lx=.f.; the results will be written to \hyperref[subsec:omega-file]{\tt omega.dat} and - if
the desired precision has been reached - to standard output
\item[\hypertarget{istep_vsp}{\tt istep\_vsp [int 0]:}] number of time steps between two consecutive 
entries into the \texttt{vsp.dat} output file (\hyperref[subsec:vsp-file]{see below}); 
this number should probably not be too small (e.g., in the range of 100-1000)
\item[\hypertarget{istep_neoclass}{\tt istep\_neoclass [int 0]:}] time steps between two entries in the \texttt{neoclass.dat} output file
(\hyperref[subsec:neoclass-file]{see below}); 
(typical value in the range of 10)
\item[\hypertarget{istep_energy}{\tt istep\_energy [int 100]:}] time steps between two entries in the \texttt{energy.dat} output file
(\hyperref[subsec:energy-file]{see below}); 
(typical value in the range of 10)
\item[\hypertarget{istep_energy3d}{\tt istep\_energy3d [int 0]:}] time
steps between two entries in the \texttt{energy3d.dat} output file
(\hyperref[subsec:energy-file]{see below}); 
(typical value in the range of 100-1000)
%\item[\texttt{trapdiag [bool f]:}] this switch specifies whether to separate the contributions from
%trapped and passing particles in the \texttt{mom.dat} output files
\item[\texttt{read\_checkpoint [bool f]:}] this switch specifies whether to start the simulation by
reading an existing \hyperref[subsec:chpt-files]{checkpoint file} 
(i.e., doing a continuation run) or by initializing it according
to one of the options that can be specified via \texttt{init\_cond}
\item[\texttt{write\_checkpoint [bool t]:}] this switch specifies whether to write out a {\tt checkpoint} 
file at the end of the run; this allows one to do continuation runs in the future; 
existing checkpoint files in the \texttt{chptdir} directory are overwritten
\item[\texttt{istep\_schpt [int 0]:}] number of time steps between two consecutive write outs of a
security checkpoint file {\tt s\_checkpoint}; this feature should probably only be used if one is 
running on a quite unstable platform -- in this case, part of a run can be reused even if the system 
crashes by renaming the {\tt s\_checkpoint} to {\tt checkpoint}; existing security checkpoint files 
in the \texttt{chptdir} directory are constantly overwritten
\item[\texttt{istep\_g1 [int 0]:}] number of time steps between two consecutive write outs of the 
distribution function in the checkpoint-style output file {\tt g1} in {\tt chptdir}. 
The main difference compared to default checkpoint files is that new entries will be appended instead 
of overwriting the previous content. Therefore, these files become quite large and should only be 
written for special applications.
\item[\texttt{write\_h5 [bool f]:}] switches all binary output (e.g., {\tt mom} and checkpoint files) 
to hdf5; {\em however, due to current restrictions in an external library, it is not possible to use this features 
if more than GENE simulation is run in parallel - for instance, in parameters scans. Furthermore, 
linking of the {\tt FUTILS} library is obligatory.}
\item[\texttt{chpt\_read\_h5 [bool f]:}] enforces hdf5 format for all {\tt checkpoint} file input 
(can be used to, e.g., convert an hdf5 checkpoint into a binary checkpoint); 
{\em aforementioned restrictions apply}
\item[\texttt{chpt\_write\_h5 [bool f]:}] enforces hdf5 format for all {\tt checkpoint} file output 
(can be used to, e.g., convert a binary checkpoint into an hdf5 checkpoint);
{\em aforementioned restrictions apply}
\item[\texttt{chpt\_read\_hac [bool f]:}] enforces {\tt ADIOS} format for all {\tt checkpoint} file input 
(can be used to, e.g., convert a hac checkpoint into a binary checkpoint)
{\em requires {\bf H}lst-{\bf Adions}-{\bf C}eckpoint interface to be linked}
\item[\texttt{chpt\_write\_hac [bool f]:}] enforces hdf5 format for all {\tt checkpoint} file output 
(can be used to, e.g., convert a binary checkpoint into a hac checkpoint)
{\em requires {\bf H}lst-{\bf A}dios-{\bf C}eckpoint interface to be linked}
\item[\texttt{many\_chpts [bool f]:}] write a binary checkpoint file per process; to be used 
on systems with broken/deprecated MPI-I/O; {\em does not support changes in resolution, MPI mapping or 
total number of processes - thus not recommended}
\item[\hypertarget{iterdb_file}{\tt iterdb\_file [str '']:}] optional string defining a 2d 
\href{http://tokamak-profiledb.ccfe.ac.uk/}{iterdb}- or DIII-D-iterdb-style file 
which can be used to extract the physical input parameters; 
see also the description of the \hyperlink{species_nml}{\tt species} and 
\hyperlink{units_nml}{\tt units} namelists.
\item[\texttt{iterdb\_time [real -1]:}] optional real value defining the closest time step (default: first) 
to be read from the data base file being declared by \hyperlink{iterdb_file}{iterdb\_file}
\end{description}

\subsubsection{The \texttt{general} namelist}
%
{\em type of operation}
\begin{description}
\item[\hypertarget{comp_type}{\texttt{comp\_type [str]:}}] can be 'EV', 'IV' or 'NC' to select between eigenvalue or initial value
computations and the algebraic solver for the neoclassical equilibrium. 'EV' is only possible for linear problems, 'NC' requires to set
\hyperlink{include_f0_contr}{\tt include\_f0\_contr} (see below).
\item[\texttt{nonlinear [bool]:}] this switch specifies whether to do a nonlinear run (if set to
\texttt{.t.}) or a linear one (if set to \texttt{.f.}); since nonlinear runs require many more
modes (which are all coupled) than linear runs, they are much more expensive
\item[\hypertarget{perf_vec}{\tt perf\_vec [int 0 0 0 0 0 0 0 0]:}] the nine-element 
{\em performance} {\em vector}\\ which defines the routines to be used for the simulation. 
Zero valued elements allow for choosing the best alternative for the corresponding code part 
automatically while positive numbers fix the set-up. As the {\tt perf\_vec} is correlated 
with the chosen parallelization, both should optimally only be fixed at once.\\
The actually chosen {\tt perf\_vec} value is printed in the output {\tt para\-me\-ters.dat} file and 
can thus be taken for continuation runs.
\end{description}
%
{\em settings for initial value runs}
\begin{description}
\item[\texttt{ntimesteps [int 1e9]:}] upper limit for the number of time steps to be computed (optional)
\item[\texttt{timelim [int]:}] wall clock time limit for the \gene computation in seconds; this number
  should be set to a value which is slightly below the time limit of your batch queue (or interactive
  node) in order to make \gene stop in a  controlled way (write checkpoint files etc.) even if
  \texttt{ntimesteps} has not been reached yet
\item[\texttt{simtimelim [real]:}] \gene is stopped in a controlled way if the simulation time (in units of
 $L_\rf/c_\rf$) exceeds this value (optional)
\item[\hypertarget{omega_prec}{\tt omega\_prec [real 1e-3]:}] 
  desired precision of growth rate and frequency for linear computations, the simulation is stopped once
  this precision is reached. Only considered if \hyperlink{istep_omega}{\tt istep\_omega} is greater than zero
\item[{\tt overflow\_limit [real 1e30]:}] defines the overflow limit with respect to the maximum 
  value of \hyperref[subsec:nrg-file]{nrg} moments
\item[{\tt underflow\_limit [real 1e-8]:}] defines the underflow limit with respect to the minimum 
  value of \hyperref[subsec:nrg-file]{nrg} moments (important if no microinstability is present)
\item[\texttt{dt\_max [real]:}] (maximum) time step to be used in the simulation; this value must,
  of course, conform to the linear stability condition; an optimized value can be determined by an
  eigenvalue computation \\(\texttt{which\_ev='largest\_magnitude'}), an approximate value is suggested by the \gene
  standard output otherwise; note that in linear runs, the time step is always kept constant, 
  while in nonlinear runs, it is reduced if the nonlinearities demand this
\item[\texttt{calc\_dt [bool f]:}] switches on automatic determination of dt\_max. If SLEPc is installed, 
  this determination is exact (sometimes \hyperlink{n_ev}{n\_ev} has to be increased,
  though), without SLEPc, dt\_max is determined from some analytical approximations
\item[\texttt{courant [real 1.25]:}] factor multiplying the estimate for the maximum time step due to 
  the ExB advection in nonlinear simulations; can be lowered if numerical instabilities are observed in the
  time trace of nonlinear runs; A value smaller than 0.95 should never be necessary. Some cases are even stable with courant=2.
\item[\texttt{timescheme [str 'RK4']:}] time scheme used for initial value calculations. Choices are 
  'RK3', 'RK4', 'RK4M', (and 'IE1p' if PETSc is available, 'IE1s' if SCALAPACK is available), 
  where RK3 and RK4 are 3rd and 4th order Runge-Kutta schemes 
  with 3 and 4 stages respectively, and RK4M is a 4th order Runge-Kutta scheme with 6 stages, where the 
  additional stages have  been used to obtain an almost optimal stability range along the imaginary axis.
  (IE1p/s are fully implicit 1st order Euler scheme relying on PETSc/SCALAPACK; 
  it is implemented for linear runs only, the numerical 
  errors are of the growth rate and frequency are corrected if istep\_omega is sufficiently low)
\item[\texttt{coll\_split\_scheme [str 'RKCa']:}] time scheme used for
  collisions with operator splitting. Choices are 'EE1' for explicit
  Euler, 'RKC2', 'RKC3', 'RKC4', 'RKCa' for 2-4 stage Runge-Kutta-Chebychev schemes with
  optimally extended stability along the negative real axis. 'RKCa' is
  adaptive. Set to 'none' to include collisions in the regular
  timescheme without operator splitting.
\end{description}
%
{\em settings for eigenvalue runs}
\begin{description}
\item[\hypertarget{ev\_prec}{\tt ev\_prec [real 1e-3]:}] 
  desired precision of growth rate and frequency
\item[\texttt{which\_ev [str 'jd']:}] can only be used if external eigenvalue solver is installed, valid choices are
  'largest\_real', 'smallest\_real', 'largest\_magnitude', 
  'shift\_invert', 'harmonic', 'gd' or 'jd' (from SLEPc 3.1) for finding
  the eigenvalues closest to ev\_shift; for solvers that use
  additive Schwartz preconditioning (like 'jd') the
  \hyperlink{parallelization_nml}{{\tt parallelization}} namelist
  settings \\{\tt n\_procs\_z}$\leq${\tt
  nz0}/12 {\tt n\_procs\_v}$\leq${\tt nv0}/12 (or -1 for automatic) are
  recommended;\\
  additionally, two direct solvers 'all\_lapack' (not parallelized!) or 'all\_mpl'
  (only with SCALAPACK) to compute all eigenvalues;
\item[\hypertarget{n_ev}{\texttt{n\_ev [int 1]:}}] number of eigenvalues to be computed
\item[\texttt{ev\_max\_it [int]:}] maximum number of iterations used in SLEPc eigenvalue computation, 
  if no value is specified the SLEPc default value is used
\item[\texttt{ev\_shift [cmplx (10.0,0.0)]:}] shift in the complex plane used for the shift and invert spectral transform
\end{description}
%
{\em settings for neoclassical computations}
\begin{description}
\item[\hypertarget{include_f0_contr}{\tt include\_f0\_contr [bool F]:}] include the f0 contribution
(the neoclassical drive term) in the kinetic equation. This is required for neoclassical runs, regardless
if \hyperlink{comp_type}{\tt comp\_type}='IV' or 'NC' is chosen. In order to separate the neoclassical equilibrium from plasma instabilities
one choses {\tt ky0\_ind}=0 and {\tt nx0}=1. In 'IV' neoclassical runs, one must ensure a sufficiently long
simulation time for the equilibrium to be found, and set \hyperlink{istep_neoclass}{{\tt istep\_neoclass}$>0$}. A timestep estimate can be precomputed with {calc\_dt}.

\end{description}
%
{\em initialization}
\begin{description}
\item[\hypertarget{init_cond}{\tt init\_cond [str 'ppj']:}] this switch specifies the initialization of the distribution
  function (for \texttt{read\_checkpoint=.f.}); one can choose from the following options:
  \texttt{'sw'} ('start wave') corresponds to a single $k_x$-$k_y$ mode,
  \texttt{'fb'} ('Fourier blob') to a Gaussian distribution in $k_x$, $k_y$, and $z$,
  \texttt{'alm'} ('all modes') to an excitation of all $k_x$ and $k_y$ modes (using powers of the
  Jacobian in the parallel direction); 
  \texttt{'gam'} ('GAM') to initialize $\langle\phi\rangle_{\rm FS} = 1$;
  \texttt{'zero'} to initialize zero fluctuations;
  besides that, there are also the options \texttt{'ppg'},
  \texttt{'ppj'}, and \texttt{'mmj'}
 % The \texttt{p, m, j} triples indicate \texttt{g}aussian distributions,
 % \texttt{p}ower laws or powers of the \texttt{j}acobian in the xyz coordinates
\item[\texttt{init\_aux\_x [real]:}] optional control parameter in the context of setting up the initial
  condition in the $x$ direction; its meaning and default value depends on the choice of \texttt{init\_cond}
  (e.g., mode number, power law exponent)
\item[\texttt{init\_aux\_y [real]:}] same as above, but for the $y$ direction
\item[\texttt{init\_aux\_z [real]:}] same as above, but for the $z$ direction
\end{description}
%
{\em species independent physical input parameters}
\begin{description}
\item[\hypertarget{beta}{\tt beta [real]:}] plasma beta as defined by $\beta_\rf=8\pi
  n_{\mathrm{e}0}T_\rf/B_\rf^2$ in cgs; respectively 
  $\beta_\rf\approx 403\cdot 10^{-5} n_{e19} T_{\rf ,\mathrm{keV}}/B_{\rf ,T}^2$
(electron density $n_{e19}$ in $10^{19}\frac{1}{m^3}$, $T_{\rf ,\mathrm{keV}}$ in units of $keV$ and 
$B_\rf$ in Tesla); set to -1 for consistent evaluation from reference values if all of them are given 
in the \hyperlink{units_nml}{\tt units} namelist;\\
 if {\tt beta} has a finite value, magnetic fluctuations are included
  in the calculation, independent of the choice of the geometry (and the $\alpha$ parameter); note
  that often, a small but finite $\beta$ value can help to increase the time step significantly (this
  is due to the fact that for $\beta\rightarrow 0$, the shear Alfv\'en wave transitions into a
  high-frequency electrostatic wave which needs to be resolved in time)
%comparison: GENE-GS2
%$\beta^{GS2} = (T_i/T_\rf) \beta^{GENE}$
\item[\texttt{collision\_op [str 'none']:}] Collision operator, can be 'pitch-angle' or 'landau' for Landau-Boltzmann
\item[\texttt{coll\_cons\_model [str 'default']:}] model for the back-reaction term of the collision operator conserving energy and momentum,
can be 'self\_adj' for a self-adjoint form, 'default'
(='xu\_rosenbluth', a little faster) or 'none' (not recommended)
\item[\hypertarget{coll}{\tt coll [real]:}] Normalized collision
frequency defined in Gaussian units as $$\nu_c=\pi \ln\Lambda\; e^4 n_\rf L_\rf/\left(2^{3/2}T_\rf^2\right)$$ which can be simplified to
$\nu_c = 2.3031\cdot 10^{-5}\left(\ln \Lambda L_\rf n_{\rf}\right)/\left(T_{\rf}^2\right)$ 
with $L_\rf[m]$, $n_{\rf}[10^{19}/m^3]$ and $T_{\rf}[keV]$ as in the \hyperlink{units_nml}{\tt units} namelist; 
set {\tt coll} = -1 for consistent evaluation from reference values if all of
them are given in the \hyperlink{units_nml}{\tt units} namelist. \\
Following Ref.~\cite{nrlformulary}, the Coulomb logarithm can be expressed as
$\ln\Lambda=24.-\ln(\sqrt{n_{\rf}\cdot10^{13}}/(10^3 T_{\rf}))$, which is
about 10 to 15, usually;\\
For reference, the conversion to the electron-ion collision rate of Hinton and
Hazeltine\cite{HintonHazeltine1976} [their Eqs.~(4.12) and (4.36)] is given by 
\begin{align*}
 \nu_{ei}(v)
 &=\frac{2^{1/2}\pi
 Z^2e^4n_i\ln\Lambda}{m_e^{1/2}T_e^{3/2}}\frac{v_{Te}^3}{v^3}
 =4Z^2\frac{n_i}{n_{\rf}}\frac{T^2_{\rf}}{T^2_e}\frac{v_{Te}^3}{v^3}\frac{v_{te}}{L_{\rf}}\nu_c
\end{align*}
with the thermal velocity $v_{te}=\sqrt{T_e/m_e}$ and the ion charge $Z$. Note: the thermal collision rate is obtained for
$v=v_{Te}=\sqrt{2T_e/m_e}$, i.e.~differing from $v_{te}$ by a factor of $\sqrt{2}$.\\
The dimensionless collisionality of the main ion species $\nu^*_i = \frac{\sqrt{2} a B_0}{B_{p0} v_{T,i} \epsilon^{3/2} \tau_i}$ (ion version of 
Eq.~6.87 in Ref.~\cite{HintonHazeltine1976}) can be expressed in terms of \gene parameters for a circular equilibrium as:
\begin{align*}
\nu^*_i = \frac{8}{3\sqrt{\pi}} \frac{q
Z^4}{\epsilon^{3/2}}\frac{R}{L_{\rf}}\frac{n_i}{n_{\rf}}\frac{T_{\rf}^2}{T^2_i}   \nu_c
\end{align*}
with $R/L_{\rf} = $\hyperlink{major\_R}{\tt major\_R}, $\epsilon =$
\hyperlink{trpeps}{\tt trpeps} and
$q=$\hyperlink{q0}{\tt q0} from the \hyperlink{geometry_nml}{\tt
geometry} namelist and $Z$ is the ion charge.\\
In a similar fashion, we obtain the electron version
\begin{align*}
\nu^*_e = \frac{16}{3\sqrt{\pi}} \frac{q
Z^2}{\epsilon^{3/2}}\frac{R}{L_{\rf}}\frac{n_i}{n_{\rf}}\frac{T_{\rf}^2}{T^2_e}   \nu_c
\end{align*}
from Ref.~\cite{HintonHazeltine1976}.
%OLD GENE version:
%Related to $L_\rf, \;n_e,\; T_e,\; m_i$ by coll $=0.003948\frac{\log \Lambda L_\rf n_e \sqrt{A_i}}{T_e^2}$
%($L_\rf,\;n_e$ and $T_e$ in units of $m,\;10^{19}\frac{1}{m^3},\;keV$; $\log \Lambda$ is the 
%Coulomb logarithm and $A_i=m_i/m_p$)
%% comparison: GENE-GS2 (collision_op = 'pitch-angle')
%vnewk(electrons)$= 2^{\frac{3}{2}} (n_e/n_\rf)(a/L_\rf)(T_\rf/T_e)^\frac{3}{2}(m_\rf/m_e)^\frac{1}{2}$ coll 
\item[\hypertarget{zeff}{\tt zeff [real 1.0]:}] effective ion charge 
\texttt{zeff}=$n_e^{-1}\sum_i n_i q_i^2$ (only used in collisions);
set to -1 to read from iterdb file with \hyperlink{prof_type}{\texttt{prof\_type}}=-2 or -3;
values of \texttt{zeff}$\ne 1$ are only valid, if no active impurity ion
species is included; note that the effect of impirities can only
partially be modelled by {\tt zeff}
\item[\texttt{debye2 [real 0.0]}] squared debye wavelength normalized to $\rho_\rf$\\
 debye2 $=\left(\lambda_{db}/\rho_\rf\right)^2=5.2936\cdot 10^{-4}\frac{B_\rf^2}{n_{\rf ,19}}\frac{m_p}{m_\rf}$\\
 ($n_{\rf ,19}$ in $10^{19}\frac{1}{m^3}$, $B_\rf$ in Tesla);\\
set to -1 for consistent evaluation from reference values if all of them are given 
in the \hyperlink{units_nml}{\tt units} namelist
\end{description}
%
{\em hyper diffusion settings}
\begin{description}
\item[\texttt{hyp\_x [real 0.0]:}] this parameter specifies the strength of the numerical dissipation
  (hyperdiffusion) in the radial direction (see also \hyperlink{hyp-advice}{FAQ}); by default, it is set to zero
\item[\texttt{hyp\_x\_order [int 4]:}] exponent of the hyperdiffusion operator in the radial direction;
  any even number is allowed
\item[\texttt{hyp\_y [real 0.0]:}] this parameter specifies the strength of the numerical dissipation
  (hyperdiffusion) in the binormal direction (see also \hyperlink{hyp-advice}{FAQ}); by default, it is set to zero
\item[\texttt{hyp\_y\_order [int 4]:}] exponent of the hyperdiffusion operator in the binormal direction;
  any even number is allowed
\item[\texttt{hyp\_z [real 0.0]:}] this parameter specifies the strength of the numerical dissipation
  (hyperdiffusion) in the parallel direction (see also \hyperlink{hyp-advice}{FAQ}); by default, it is set to zero
\item[\texttt{hyp\_z\_order [int 4]:}] exponent $n$ of the hyperdiffusion operator $D_n$ in the parallel
  direction; the latter is defined by $D_n = - i^n\,\texttt{hyp\_z}\,(\Delta z/2)^n\,(d/dz)^n$; one may
  choose between $n=2$ and $n=4$
\item[\texttt{hyp\_v [real 0.0]:}] this parameter specifies the strength of the numerical dissipation
  (fourth-order hyperdiffusion) in the parallel velocity coordinate (see also \hyperlink{hyp-advice}{FAQ}); 
by default, it is set to zero
\end{description}
%
{\em miscellaneous settings}
\begin{description}
\item[\hypertarget{bpar}{\tt bpar [bool f]:}] switches on/off the consideration of parallel magnetic fluctuations if 
  \hyperlink{beta}{\tt beta} is larger than zero by setting it to \texttt{.t.}/\texttt{.f.}
%\item [\texttt{pressure\_term [bool f]:}] include pressure gradient contribution in the curvature drift if set to {\tt .true.}.
%By default, this term is ignored as it is of the same order as $B_\|$ terms which are only considered if {\tt bpar} is set 
%to true. The pressure gradient amplitude is either taken from the geometry interface or computed self-consistently from 
%density and temperature gradients.
\item[\texttt{delzonal [bool f]:}] this parameter may be used to
artificially suppress all zonal components of $\phi$ (zonal flows)
  by setting it to \texttt{.t.}
\item[\texttt{delzonal\_fields [bool f]:}] this parameter may be used to
artificially suppress all zonal components of $A_{\parallel}$ (zonal
fields) by setting it to \texttt{.t.}
\item[\texttt{add\_zonal\_phi [real 0.0]:}] this parameter may be used to define the amplitude 
of an artificial zonal component which is added to the electrostatic potential at the end of each
field solve. In combination with {\tt init\_cond = 'zero'}, this may for instance be used to study 
residual zonal flow levels as predicted by Rosenbluth/Hinton and others.
\item[\texttt{avgflux\_stime [real -1.0]:}] activates computation of time and volume averaged fluxes starting 
at time \texttt{avgflux\_stime} if set to 0.0 or larger; this diagnostic is called within the \texttt{nrg} 
diagnostic, i.e. every \texttt{istep\_nrg} time step.
% \item[\texttt{parscheme [str 'c4th']:}] this parameter defines the finite differencing scheme in the
%   parallel direction; one can choose between \texttt{'c4th'} (fourth-order centered), \texttt{'c2nd'}
%   (second-order centered), and \texttt{'u3rd'} (third-order upwind); under normal circumstances, the
%   first of these options is recommended.
\item[\texttt{perf\_tsteps [int 3]:}] defines the number of simulation time steps to be computed for the 
time measurements in the automatic \hyperlink{perf_vec}{\tt perf\_vec} determination.
\end{description}
%
\newpage
{\noindent\em antenna drive settings}
\begin{description}
\item[\hypertarget{antenna_type}{\tt antenna\_type [int 0]:}] Determines the type of antenna used to 
drive externally imposed currents. Four settings are possible: 
\begin{itemize}
\item 0 -- antenna off
\item 1 -- sinusoidal antenna applied to all modes
\item 2 -- Langevin antenna, included in the modified distribution function $g_{1j}$
\item 3 -- Langevin antenna, not included in $g_{1j}$, with explicit $\partial A_{1\|}/\partial t$ term (recommended)
\end{itemize}
\item[\hypertarget{Apar0_antenna}{\tt Apar0\_antenna [real 0.0]:}] Amplitude for type 1 antenna.
\item[\hypertarget{omega0_antenna}{\tt omega0\_antenna [real 0.0]:}] Frequency for type 1 antenna.
\item[\hypertarget{lv_antenna_modes}{\tt lv\_antenna\_modes [integer*N*3 (0,0,0)*N]:}] Defines the 
modes driven by the Langevin antenna (see TenBarge et al. (CPC 2014). For each mode tuple ($k_x$,$k_y$,$k_z$), 
the amplitude is evolved using a separate Langevin equation. Specify modes in the form 
$i_1$,$j_1$,$k_1$,$i_2$,$j_2$,$k_2$ etc., where $i$, $j$, $k$ are the indices of the $k_x$, $k_y$, $k_z$ modes. 
\item[\hypertarget{lv_antenna_amp}{\tt lv\_antenna\_amp [complex*N (0,0)]:}] Sets the amplitude for 
Langevin-type antenna drive. Amplitudes must be given for each mode tuple. For continuation runs, it is 
necessary to use the amplitude from the output parameters file (amplitude at the end of the last simulation) 
to enable a smooth continuation. In this case, the parameter {\tt lv\_antenna\_initamp} {\em must} be 
specified to allow for an equal treatment of the random contribution to the Langevin equation. 
\item[\hypertarget{lv_antenna_initamp}{\tt lv\_antenna\_initamp [complex*N (0,0)]:}] Specifies the initial 
(=target) amplitude for Langevin-type antenna drive. This parameter is only required for continuation runs in order 
to allow a consistent setup of the random number contribution to the Langevin equation.
\item[\hypertarget{lv_antenna_freq}{\tt lv\_antenna\_freq [complex*N (0,0)]:}] Defines the frequency and 
decorrelation rates for each Langevin antenna. Decorrelation rates have to be explicitly specified as 
negative values, otherwise the antenna amplitudes will grow exponentially. 
\end{description}

\subsubsection{The \texttt{external\_contr} namelist}
This namelist is used to enter externally imposed equilibrium flows and stationary temperature and density
profiles.
\begin{description}
\item[\texttt{ExBrate [real 0.0]:}] This parameter defines a radially constant ExB shearing rate
  (determined by toroidal rotation), which is implemented by shifting the radial Fourier mode grid in time
  (see Ref. \cite{Hammett06}). The parameter is given in normalized units by 
  \texttt{ExBrate}$=-\frac{x_0}{q_0}\frac{\partial\Omega_{\mathrm{tor}}}{\partial x}\frac{L_\mathrm{ref}}{c_\mathrm{ref}}$,
  where $x_0$ is the radial position in the \gene coordinate system ($\rho_\mathrm{tor}$ for realistic geometry,
  $r/L_\mathrm{ref}$ for 
  circular/s-$\alpha$ geometry), $q_0$ is the safety factor and $\partial\Omega_{\mathrm{tor}}/\partial x$ 
  is the radial derivative of the 
  toroidal angular velocity. A realistic value for a conventional tokamak would be e.g. \texttt{ExBrate}$=0.05$;\\
  set to -1111.0 for consistent evaluation from reference values if all of them are given 
  in the \hyperlink{units_nml}{\tt units} namelist and an \hyperlink{iterdb_file}{iterdb\_file} containing \texttt{VROT} 
  has been specified.
\item[\texttt{pfsrate [real 0.0]:}] This parameter defines a shearing rate to model the parallel flow shear
  drive that appears due to toroidal rotation. Typically, it is used in combination with finite \texttt{ExBrate} and 
  defined in the same way (additional geometric factors are added automatically by \gene). 
  Hence, the recommended setting for this parameter is \texttt{pfsrate}$=-1111$, which will set the parallel flow shear
  equal to the \texttt{ExBrate} as it is given for a purely toroidal flow.
\item[\texttt{Omega0\_tor [real 0.0]:}] Angular toroidal rotation in units of $c_\rf/L_\rf$; if set to finite values, simulations will be run in an accordingly rotating co-moving frame with Coriolis, Centrifugal and further comoving frame effects activated unless explicitly altered by {\tt with\_coriolis}, {\tt with\_centrifugal}, and {\tt with\_comoving\_other}; use -1111.0 for consistent evaluation from reference value \hyperlink{omegatorref}{\tt omegatorref} if the latter is provided in the \hyperlink{units_nml}{\tt units} namelist
\item[\texttt{ExB\_stime [real 0.0]:}] With this parameter, the user can set the time (in simulation units)
  when ExB flow shear is activated - for instance, to a time where the unsheared turbulence has reached a quasi-stationary 
  state. This is especially recommended for strong flow shear which would otherwise
  often lead to very long saturation times or an initial suppression of turbulence. 
  Note that parallel flow shear is not affected by this parameter and will be active from the beginning of the simulation.
\item[\texttt{<var>0\_ext [real 0.0]:}] These parameters specify the amplitude for sinusoidal contributions to
  the electrostatic potential, the temperature profile or the density profile (set \texttt{<var>} to \texttt{phi}, 
  \texttt{omt}, \texttt{omn} or \texttt{apar}, respectively).
\item[\texttt{kxind\_<var>\_ext [int -1]:}] These parameters define the wavenumbers of the sinusoidal 
  contributions activated by the \texttt{<var>0\_ext} parameters. The possible definitions of \texttt{<var>}
  are given above.
\item[\texttt{phase\_<var>\_ext [int -1]:}] These parameters define radial shifts (in rad) of the sinusoidal 
  contributions activated by the \texttt{<var>0\_ext} parameters. See above for possible definitions of \texttt{<var>}.
\end{description}
{\em Optional switches for detailed comoving frame effect studies:}
\begin{description}
\item[\texttt{with\_coriolis [bool]:}] Switch on/off Coriolis effects in the Vlasov equation; default is true for finite {\tt Omega0\_tor} and false else
\item[\texttt{with\_centrifugal [bool]:}] Switch on/off Centrifugal effects in the Vlasov equation; default is true for finite {\tt Omega0\_tor} and false else
\item[\texttt{with\_comoving\_other [bool]:}] Switch on/off all implemented comoving frame effects beyond Coriolis and Centrifugal forces, e.g.~the formation of an equilibrium potential in combination with poloidal equilibrium density assymetries and the modification of the trapping and linear drive terms; default is true for finite {\tt Omega0\_tor} and false else\\
{\bf note:} a reference position for the density normalization has to be defined via {\tt R0\_tor} if poloidal assymetries are considered!
\item[\texttt{R0\_tor [real -1]:}] Reference radial position for the equilibrium density normalization if poloidal assymetries are taken into account; set to magnetic axis (-1), low field side radius (-2) or specify positive valued position by hand
\item[\texttt{dR0\_tor\_dx [real]:}] Radial derivative of {\tt R0\_tor} if the latter is positive valued
\end{description}




\subsubsection{The \texttt{geometry} namelist}
\hypertarget{geometry_nml}{}
\begin{description}
\item[\hypertarget{magn_geometry}{\tt magn\_geometry [str]:}] this parameter defines the magnetic geometry used in the computation. 
  The analytical models supported are 'slab' (basic slab model), 'slab\_curv' (slab with curvature), 
  's\_alpha', 'circular' (circular concentric flux surface; $x = r$), and 'miller'; 
  to use the geometric information from 
  (numerical) MHD equilibria choose 'chease', 'tracer\_efit' or 'gist';
  \gene geometry output files being defined by \texttt{geomfile} can be reread 
  by choosing 'gene' (starting from release 1.6)
%
%
\item[{\parbox[t]{0.9\linewidth}{additional parameters for analytical equilibrium models\\
 's\_alpha', 'circular', and 'miller':}}]
\item[\hypertarget{q0}{\texttt{q0 [real]:}}] safety factor $q$ at the central radial position of the flux tube
\item[\texttt{shat [real 0.0]:}] magnetic shear parameter as defined by $\hat s=(\rho/q)(dq/d\rho)$ with flux surface label $\rho$ 
 ($\rho = r$ for most models); important for parallel boundary condition; the value is overwritten in case of numerical equilibria
\item[\texttt{amhd [real 0.0]:}] $\alpha$ parameter as used in magnetohydrodynamics and defined by
  $\alpha_{\rm MHD}=-q^2R\,(d\beta/dr)$ where $\beta=8\pi p/B_\rf^2$; for \texttt{amhd=-1}, it is set automatically
  such that it is consistent with $\beta$ and the temperature and density gradients; otherwise, any
  other value can be chosen in order to study Shafranov shift effects in the geometry independently of
  {\em dynamical} finite $\beta$ effects (associated with magnetic field fluctuations); note: 'circular' always assumes 
  concentric ($\alpha_{\rm MHD} = 0$) flux surfaces
\item[\texttt{major\_R [real 1.0]:}] major radius ($R/L_\rf$) of the device--for \hyperlink{desc:miller}{Miller} geometry, 
of the flux surface--in units of the macroscopic length scale $L_\rf$
\item[\hypertarget{minor_r}{\texttt{minor\_r:}}] minor radius of the device in units of the macroscopic length scale ($a/L_\rf$); 
 only required for 'circular' and 'miller'
\item[\hypertarget{trpeps}{\texttt{trpeps [real 0.0]}}:] inverse aspect ratio at the flux tube position, $\epsilon=r/R$
%
\item[additional parameters for \hypertarget{desc:miller}{'miller'}:]
%\item[\texttt{rho [real]:}] minor radius $r$ of the flux tube in units if the macroscopic reference length ($r/L_\rf$)
\item[\texttt{kappa [real 1.0]:}] elongation $\kappa$
\item[\texttt{delta [real 0.0]:}] triangularity $\delta$
\item[\texttt{zeta [real 0.0]:}] squareness $\zeta$
\item[\texttt{s\_kappa [real 0.0]:}] $s_\kappa = \frac{r}{\kappa} \pderiv{\kappa}{r}$
\item[\texttt{s\_delta [real 0.0]:}] $s_\delta = \frac{r}{\sqrt{1-\delta^2}} \pderiv{\delta}{r}$; warning: other codes use $r \pderiv{\delta}{r}$ instead
\item[\texttt{s\_zeta [real 0.0]:}]  $s_\zeta = r \pderiv{\zeta}{r}$
\item[\texttt{drR [real 0.0]:}] shift of the major radius of the flux surface $\pderiv{R}{r}$
\end{description}
{\em Note:} The GENE implementation of Miller geometry uses $x=r$ as the radial coordinate. 
Thus, all input gradients (including {\tt amhd}, {\tt shat}, {\tt ExBrate}) are expected to be 
given in terms of $\frac{d}{dr}$ derivatives. A Python script for an automatic determination of 
the Miller parameters from a G-EQDSK (EFIT) file can be found in {\tt ./tools/python/extract\_miller\_from\_eqdsk.py}. 
The script will write a block of input parameters to standard output, which can be copied directly 
to a GENE parameters file. Note that presently, automatic profile input from files is not possible for Miller geometry.
Contrary to most geometry interfaces, the Miller $B_\rf$ is not taken on axis but at the center 
of the flux surface (typically the difference is quite small).
%
\begin{description}
\item[additional parameters for 'slab', 'slab\_curv':]
\item[\hypertarget{q0}{\texttt{q0 [real]:}}] safety factor $q$ at the central radial position of the flux tube
\item[\texttt{shat [real 0.0]:}] magnetic shear parameter as defined by $\hat s=(\rho/q)(dq/d\rho)$ with flux surface label $\rho$ 
for sheared slab applications; important for parallel boundary condition
\item[\texttt{parscale [real 1.0]:}] defines the parallel scale length
in units of $L_\rf$ for slab geometry\\
\textbf{note}: (a) this parameter enters only into the Jacobian and - in case of 'slab\_curv' in the curvature drift; (b) for
comparisons with \cite{diffusion} (diffusion study of slab ITG), one needs to
rescale the gradient lengths and the growth rate with $\alpha_\mathrm{i} / \sqrt{2}$;
(c) set {\tt parscale} to {\tt q0} for direct comparison with 's\_alpha' at {\tt trpeps}$\ll 1$ or renormalize results otherwise
%
\item[additional parameters for 'tracer', 'tracer\_efit' or 'gist':]
\item[\texttt{geomdir [str './']:}] directory from which the geometry file is read
\item[\texttt{geomfile [str]}:] name of the geometry file; note that for 'tracer' a file generated by
the TRACER code is required, whereas 'tracer\_efit' will read an EFIT file, which will then be traced
by a \textsc{Gene}-internal version of TRACER.
\item[additional parameters for 'chease':]
\item[\texttt{geomfile [str]}:] name of the geometry file (in hdf5 format), if
  no file name is given, the code looks for a file called ogyropsi.h5.
\item[\texttt{x\_def [str 'arho\_t']:}] Select definition of x variable.  It can be
  set to 'arho\_p' $=a \sqrt{\Psi_p / \Psi_{p\mbox{ edge}}}$ where $\Psi_p$ is the
  poloidal flux, 'arho\_t' $=a \sqrt{\Psi_t / \Psi_{t\mbox{ edge}}}$ where $\Psi\_t$ is the
  toroidal flux or 'arho\_v' $=a \sqrt{V / V_{\mbox{ edge}}}$ where $V$ is the
  volume inside a flux surface, a is the minor radius. It can also be set to 'C\_psi'
  $=q0/(r_0 B_0) \Psi_p$ where $r_0$ and $q_0$ are the local minor radius and
  safety factor, and $B_0$ the magnetic field at the axis. The choice of x\_def, should be considered
  when setting omn=$-L_{\rf} \, d \ln n/dx$ and omt=$-L_{\rf} \, d \ln T /dx$.
%\item [\texttt{flux\_pos [real 0.0]:}] position $x/a$ at which the equilibrium
%  is taken (between 0 and 1) 
\item[Miscellaneous:]
\item[\hypertarget{dpdx_pm}{\tt dpdx\_pm [real]:}] amplitude of the {\em negative} equilibrium pressure gradient $-\nabla p/p_m$ 
normalized to the reference magnetic pressure $p_m = 2\mu_0/B_\rf^2$. Set to 
\begin{itemize}
 \item[$\geq 0$] for user-defined value
 \item[$-1$] for automatic computation from density and temperature gradients via
 \texttt{dpdx\_pm} $=\beta_\rf \sum_\spec \texttt{dens}_\spec {\tt temp}_\spec (\frac{L_\rf}{L_{T_\spec}}+\frac{L_\rf}{L_{n_\spec}})$
 \item[$-2$] for automatic computation from MHD equilibrium ({\em recommended} if available!)
\end{itemize}

\item[\hypertarget{dpdx_term}{\tt dpdx\_term [str]:}] defines the way how the pressure gradient is considered or neglected in the 
combined curvature and $\nabla B$ drift $\mvec{v}_{{\rm curv}+\nabla B}$
\begin{itemize}
 \item {\tt 'full\_drift':} ({\em recommended} for simulations with $B_\|$ fluctuations); consider full drift velocity
 \bea
 \mvec{v}_{{\rm curv}+\nabla B} & = \mvec{b}\times\left[\left(v_\|^2+\frac{v_\perp^2}{2}\right)\frac{\nabla B}{B}+v_\|^2\frac{4\pi}{B^2}\nabla p\right]/\Omega \nn \\
 & = \mvec{b}\times\left[\left(v_\|^2+\frac{v_\perp^2}{2}\right)(\mvec{b}\cdot\nabla)\mvec{b}-\frac{v_\perp^2}{2}\frac{4\pi}{B^2}\nabla p\right]/\Omega \nn
 \eea
 requires {\tt dpdx\_pm} to be set
 \item {\tt 'gradB\_eq\_curv':} (often {\em recommended} for simulations without $B_\|$ fluctuations); erase pressure gradient from $\nabla B$ drift, i.e.
 \bea
 \mvec{v}_{{\rm curv}+\nabla B} & \sim \mvec{b}\times\left[\left(v_\|^2+\frac{v_\perp^2}{2}\right)\frac{\nabla B}{B}+\left(v_\|^2+\frac{v_\perp^2}{2}\right)\frac{4\pi}{B^2}\nabla p\right]/\Omega \nn
 \eea
 in the $\nabla B/B$ formulation employed in {\tt Gene}; requires {\tt dpdx\_pm} to be set
 \item {\tt 'curv\_eq\_gradB':} erase pressure gradient from curvature drift (formerly the default in \gene if {\tt pressure\_term = F} had been set)
 \bea
 \mvec{v}_{{\rm curv}+\nabla B} & \sim \mvec{b}\times\left[\left(v_\|^2+\frac{v_\perp^2}{2}\right)\frac{\nabla B}{B}\right]/\Omega \nn
 \eea
 {\tt dpdx\_pm} is not required
 \end{itemize}

\item[Optional:]

\item[\hypertarget{norm_flux_projection}{\tt norm\_flux\_projection [bool f]:}] Use normalized/unnormalized radial projections 
when computing radial fluxes in the {\tt nrg} file (see Sec.~\ref{subsec:nrg-file}).

\item[\hypertarget{rhostar}{\tt rhostar [real]:}] defined as $\rho^* = \rho_\rf/a$ with minor radius $a$ being set by {\tt minor\_r} 
or the numerical equilibrium interfaces; set to -1 for consistent evaluation from reference values if all of them are given 
in the \hyperlink{units_nml}{\tt units} namelist; note that {\tt rhostar} is not required in fluxtube simulations unless certain 
parameters like \hyperlink{n0\_global}{\tt n0\_global} shall be set consistently w.r.t.~$\rho_\rf$ and $a$.\\

\end{description}

\subsubsection{{The \texttt{species} namelist}}
\hypertarget{species_nml}{}
\begin{description}
\item[\texttt{name [str]:}] name of the particle species in the present namelist (may be chosen freely)
\item[\texttt{passive [bool f]:}] if this parameter is set to \texttt{.t.}, the particle species is
only treated as a passive tracer, i.e., no fields generated by this species are computed and 
there are no back-reactions on the turbulence
\item[\texttt{omn [real 0.0]:}] normalized density gradient of the present particle species,
  $\omega_n=-(L_\rf/n)\,(dn/dx)$; if one chooses \texttt{major\_R}=1.0, one has $\omega_n=R/L_n$
\item[\texttt{omt [real 0.0]:}] normalized temperature gradient of the present particle species,
  $\omega_T=-(L_\rf/T)\,(dT/dx)$; if one chooses \texttt{major\_R}=1.0, one has $\omega_T=R/L_T$
\item[\texttt{mass [real]:}] mass of the present particle species, normalized to an arbitrary mass scale $m_\rf$
\item[\texttt{charge [int]:}] signed charge of the present particle species, normalized to the elementary charge
  (electrons: \texttt{charge=-1})
\item[\texttt{temp [real 1.0]:}] temperature of the present particle species, normalized to an arbitrary temperature
scale $T_\rf$
\item[\texttt{dens [real 1.0]:}] density of the present particle species, normalized to the electron density;
in order to ensure quasineutrality, the condition $\sum \texttt{charge}\cdot\texttt{dens}=0$ (sum over all
{\em active} species) must be satisfied
\item[\hypertarget{prof_type}{\tt prof\_type [integer 0]:}] the following options are available:
\begin{itemize}
 \item[0] use the aforementioned user-defined \texttt{omn}, \texttt{omt}, \texttt{temp} and \texttt{dens} settings; for local code only
 \item[-1] use profile data provided in a file named \texttt{profiles\_<species name>} 
  for temperature, density and their gradients; for further information on the actual file format,
  see \texttt{src/profiles.F90}, subroutine \texttt{read\_tempdens\_profile}
 \item[-2] use profile data provided in the \hyperlink{iterdb_file}{\tt iterdb\_file} 
  for temperature, density and their gradients; the relevant entries read by \gene are {\tt NE},
  {\tt TE}, {\tt NM1} (,{\tt NM2}), {\tt TI}, {\tt VROT} and {\tt ZEFFR}; the file format is discussed
  here \url{http://tokamak-profiledb.ccfe.ac.uk/}
 \item[-3] use DIII-D iterdb style profile data provided in the
  \hyperlink{iterdb_file}{\tt iterdb\_file} for temperature, density and their gradients;
  the relevant entries read by \gene are {\tt nj}, {\tt electron temperature}, {\tt ion temperatue}
  (sic), {\tt electron density}, {\tt primary ion density} (, {\tt impurity ion density}),
  {\tt angular rotation speed}, and {\tt zeff profile}
 \item[$>0$] analytical profiles for {\em global} simulations
\end{itemize}
Be careful to use {\em consistent} radial coordinates in your profile files and the magnetic geometry. In most cases, the normalized $\rho_{\rm tor}$ is employed for this purpose.
\end{description}
For each particle species, a separate \texttt{species} namelist has to be provided; \gene will then use
the first \texttt{nspec} such namelists, ignoring others (if present).

\subsubsection{{The \texttt{units} namelist (optional)}}\label{sec:units_nml}
\hypertarget{units_nml}{}
The original purpose of this namelist was to allow post-processing tools to convert the normalized observables to SI units as those values had just been passed to the output \texttt{parameters.dat} file. However, the reference values being described in the following can now also be used to automatically compute physical input variables like \hyperlink{beta}{\tt beta} or \hyperlink{coll}{\tt coll} if {\bf all} of them are either explicity set or consistently evaluated from given profile files or interfaces (see below).
\begin{description}
\item[\texttt{Tref [real 0.0]}:] reference temperature in keV; set to -1 to automatically evaluate and consider $T_e$ at radial domain center if \hyperlink{prof_type}{\tt prof\_type} is properly set ($<0$) in {\em each} species namelist.
\item[\texttt{nref [real 0.0]}:] reference density in $10^{19}\,{\rm m}^{-3}$; set to -1 to automatically evaluate and consider $n_e$ at radial domain center if \hyperlink{prof_type}{\tt prof\_type} is properly set ($<0$) in {\em each} species namelist.
\item[\texttt{mref [real 0.0]}:] reference mass in units of the proton mass; set to -1 for deuterium.
\item[\texttt{Bref [real 0.0]}:] reference magnetic field in T; will always be filled automatically if the \hyperlink{magn_geometry}{magn\_geometry} is set to \texttt{tracer}, \texttt{tracer\_efit} or \texttt{chease}.
\item[\texttt{Lref [real 0.0]}:] reference length in m; will always be filled automatically if the \hyperlink{magn_geometry}{magn\_geometry} is set to \texttt{tracer}, \texttt{tracer\_efit} or \texttt{chease}.
\item[\hypertarget{omegatorref}{\texttt{omegatorref [real 0.0]}}:] reference toroidal angular velocity in ${\rm rad}/{\rm s}$; set to -1 to automatically evaluate from data base if \hyperlink{prof_type}{\tt prof\_type} is properly set ($<0$) in {\em each} species namelist; needed for Doppler shifted post-processing.
\end{description}



\newpage
\subsection{Running the code}
After you have edited the \texttt{parameters} file (make sure that
input files and output directories exist), you can either start
\gene interactively using mpiexec or poe (depending on your
machine) or submit it to a batch queue. In any case, you have to
specify the total number of MPI processes, which is given by
$N_\mathrm{MPI}=\mathtt{n\_procs_s}\cdot
\mathtt{n\_procs_w}\cdot\mathtt{n\_procs_v}\cdot\mathtt{n\_procs_y}\cdot\mathtt{n\_procs_z}$.
For shared memory architectures, the number of OMP processes can
be set independently by setting the environment variable
OMP\_NUM\_THREADS; note that the total number of processors needed
is then $N_\mathrm{MPI}$*OMP\_NUM\_THREADS.


\subsection{The scan script}\label{sec:scanscript}
For \gene parameter variations, use of the scan script is recommended. It can be found
in the \texttt{prob} directory.
Instructions for specification of parameter ranges are given below.
Multidimensional parameter scans are possible.
The scanscript also enables one to let parameters depend on other parameters.
For each scan a new directory named \texttt{scanfiles\#\#} in the specified
\texttt{diagdir} is created, where \texttt{\#\#} is an incremented number. If, for
example, the directory \texttt{"scanfiles0000"} exists,
\texttt{"scanfiles0001"} is created.
The output files are stored in this \texttt{scanfiles} directory. Each
parameter combination simulated is assigned a run number, which acts as a
file ending of all the corresponding output files. 
In addition, a file called \texttt{scan.log} is created at the end of
the scan, briefly summarizing the results (run number, the changed parameters, and growth rate /
frequency data). 
This (and other) data can be analyzed using the scan tab in the diagnostic tool. 
Incomplete scans may be continued by adding the \texttt{--cs} option when calling the
script (details below).
Similar to the \texttt{scan.log}, the \texttt{neo.log} file is created for neoclassical
computations, summarizing the last lines of
\hyperlink{neoclass-file}{\texttt{neoclass}} output files.

It is strongly recommended to familiarize with the \gene option of running multiple \gene
subroutines simultaneously to avoid waste of computational resources. Please read Sec.~\ref{sec:scangene} for correct use. The point is that automatic settings may lead to idling processors.

To initiate a \gene scan, one has to change the lines in the \texttt{parameters} file (or
the launcher tool) as specified below. The scanscript is executed by
replacing the last line of the submit script (the \gene call command)
by the following line 
./scanscript \texttt{--np $\langle$np$\rangle$} \texttt{--mps $\langle$mps$\rangle$} ,
where \texttt{$\langle$np$\rangle$} is the number of processes of the
MPI environment, \texttt{$\langle$mps$\rangle$} is the maximum number of
parallel scans. For more options see Sec.~\ref{sec:scanoptions}. 

\subsubsection{How to specify scan parameters:} \label{sec:scandef}

\begin{itemize}
\item $\langle$par$\rangle$ = $\langle$value$\rangle$ !scanrange: $\langle$start$\rangle$,
$\langle$step$\rangle $,$\langle$end$\rangle$

The last three values (start, step, and end) specify the interval to be scanned. The value after
the ``='' will be replaced by the actual values when running the script. In this fashion,
all parameter lines that are to be scanned over need to be adjusted.

Example:

\texttt{omn = 3.0 !scanrange: 4.0,1.0,10.0}

\item $\langle$par$\rangle$ = $\langle$value$\rangle$ !scan: $\langle$start$\rangle$,
$\langle$step$\rangle $,$\langle$end$\rangle$

Performs the same operation as !scanrange: but is not to be confused
with the following functionality.

\item $\langle$par2$\rangle$ = $\langle$value$\rangle$ !scan:
F($\langle$pars$\rangle$($\langle$\#$\rangle$))

A parameter can depend on other variable or fixed parameters.
Scanscript evaluates the function F before every \gene start and assigns
the result to the parameter $\langle$par2$\rangle$. 
After every argument of F, a bracketed number is mandatory: if the
argument parameter under consideration is part of a species namelist,
this is the
species number, otherwise it is ``1''.
This also works in functions and lists.

Example:

\texttt{omn = 3.0 !scan: omn(1)+2*omt(2)-beta(1)}

\item 
$\langle$par$\rangle$ = $\langle$value$\rangle$ !scanlist: $\langle 
$value1$\rangle$,$\langle$value2$\rangle$,$\ldots$

The scanscript also accepts comma separated lists for scans. 

Examples:

\texttt{nz0 = 1.0 !scanlist: 24,32,48}

List elements are also allowed to depend on other parameters:

\texttt{omt = 1.0 !scanlist: omn(1)+1,4,3,8}

List elements can be character strings :

\texttt{which\_ev  = 'jd' !scanlist: "'jd'","'harmonic'"}

\item $\langle$par$\rangle$ = $\langle$value$\rangle$ !scanfunc:
$\langle $maxsteps$\rangle$,$\langle$f(xi)$\rangle$,$\langle$end$\rangle$

One can use variable steps by
specifying a (strictly monotonic) function f(xi), a maximum number of steps
maxsteps, and a maximum for f(xi) above which no further steps are taken. If one of these two
conditions (exceeded number of steps or maximum value) are fulfilled, the scan
terminates at that point. The scan always starts with xi = 1 and increments 1
with every step. Possible
operators in the function are \texttt{+, -, *, /, \%, e, **, abs, sqrt, int, exp, log,
sin, cos}. Parameters can be inserted as described further down.

Example:

\texttt{omn = 3.0 !scanfunc: 10,log(xi)+omn(1),100}

One has to take care that if there are parameters in f(xi) that are changed
during the scan, the number of steps in this dimensions could vary. Also, in multidimensional
scans, if f(xi) changes with the change of another dimension, some analyses of the
\texttt{scan.log} file in the diagnostic may be affected.

Should the steps width of 1 in x not be sufficient, one can also create a dummy parameter
which can be changed arbitrarily. Please take note that the
dummy parameter's name has to start with an exclamation mark (``!'') in order to be recognized
by \gene as such. 

Example: 

\texttt{!dummy = 0 !scan: 0, 0.1, 5}

\texttt{omn = 0 !scan: omn(1)+2*omt(2)-!dummy(1)}

\end{itemize}

\subsubsection{Options (short form in brackets):}\label{sec:scanoptions}
\begin{itemize}

\item \texttt{--n\_pes=$\langle$\#$\rangle$ (--np=$\langle$\#$\rangle$)}

It is required to set $\langle$\#$\rangle$ the total number of processes
the script is started with (MPI processes $\times$ openMP processes)

\item \texttt{--procs\_per\_node =$\langle$\#$\rangle$ (--ppn=$\langle$\#$\rangle$)}

(recommended) $\langle$\#$\rangle$ is the number of processors per
compute node of your machine, used for parallel efficiency optimization.

\item \texttt{--max\_n\_par\_sims =$\langle$\#$\rangle$ (--mps=$\langle$\#$\rangle$)}

(recommended) $\langle$\#$\rangle$ is the maximum number of parallel
simulations, used for parallel efficiency optimization.

\item \texttt{--help}

Prints a short help text 

%\item \texttt{--startnum=$\langle$\#$\rangle$} 
%
%If the script was terminated prematurely, it can be restarted using this (and possibly
%the next) option. For $\langle$\#$\rangle$ the number
%of the next run is to be inserted--- if n runs exist already, n+1 has to be entered here.
%
%\item \texttt{--scandir=$\langle$path$\rangle$}
%
%If the scan data is to be written to an existing directory (e.g., when restarting
%the scan script after it was terminated), it can be specified here. By convention,
%\texttt{path} has to be relative to the path provided in the \texttt{parameters} file.

\item \texttt{--syscall}

Set the command for starting \gene, if the default does not work. This depends on the machine.

\item \texttt{--test }

Test the scanscript settings without executing \gene

\item \texttt{--force (--f)}

Continue even when errors occur (\texttt{--test} \texttt{--force} is
useful).

\item \texttt{--continue\_scan (--cs)}

If the script was terminated prematurely, it can be restarted using this option.
Make sure that the entries of the \texttt{\&scan} namelist
are correct:

\texttt{scan\_dims} is a vector giving the dimensions of the scan

\texttt{par\_in\_dir} is the subdirectory \texttt{in\_par} of \texttt{diagdir}. This directory contains all
parameter files \gene is started with.

These parameters are automatically set by the scanscript at the first
submission of the scan.

\item \texttt{--mk\_scanlog (--mks)}

Creates the scan.log from existing data in the scanfiles directory without running
\gene\\ (execute in probdir from command line without specifying
\texttt{--n\_pes}).

\item \texttt{--efficiency (--eff)}

Test the parallel efficiency for your problem. All scans are ignored.
The maximum number of processes is the value given with \texttt{--n\_pes}. \gene
autoparallelization should be used.\\
It is recommended to use the option \texttt{--procs\_per\_node}
(processes per node of the machine) and \texttt{--min\_procs} (minimum
number of processes that should be tested).\\
depending on the parameter \texttt{comp\_type}, either initial value
computations are run 10 timesteps with \texttt{dt\_max}$=1e-6$ or
a single eigenvalue solver iteration is performed to test the speed. \\ 
An efficiency.log file is created in the scan output directory. This
file contains the resulting time per step and cpu-time per step. Having
access to gnuplot, you can
plot the efficiency data using the effplot script belonging to the \gene
tools, which also creates a \texttt{.ps} file:

\texttt{effplot efficiency.log}

Note that using too many processors can not only be inefficient, but also can slow down the
computation in real time!

\item \texttt{--noeff}

Suppresses the parallel efficiency test, all relevant parameters have to
be set manually.
\end{itemize}

\subsubsection{\gene parallelization over parameter sets}

\label{sec:scangene}
\gene is able to call multiple instances of a \gene subroutine
simultaneously, so that different points in the specified parameter space can be 
computed in parallel. This is particularly useful for large parameter
scans, since parallel efficiency of a single \gene run typically
decreases with the number of processors.

For this feature, two additional parameters in the "\&parallelization"
namelist exist.

\texttt{n\_procs\_sim} is the total number of mpi processes one
instance of \gene uses

\texttt{n\_parallel\_sims} is the number of \gene subroutines running in
parallel

The product of \texttt{n\_procs\_sim} and
\texttt{n\_parallel\_sims} must equal the total number of processors of
the MPI environment (\texttt{--n\_pes}).

If you are not sure which values of \texttt{n\_procs\_sim} and
\texttt{n\_parallel\_sims} you should use, leave them blank to let the
scanscript perform a parallelization efficiency test for your problem.
In this case you must use the autoparallelization functionality by
setting some or all other entries of the \texttt{\&parallelization} namelist to $<1$.
The result is found in the efficiency.log in the output directory.
However, low dimensional parameter scans should use
\texttt{n\_parallel\_sims}=1 to make sure that processors do not idle at
the end of the scan. For the same reason, \texttt{n\_parallel\_sims}=1
is also recommended, when each single point is computationally
expensive.

A waste of resources occurs in the following example, the parameter $ky0$ is to take 9 values in the scan.
Each run performs best at a total processor number of 16. The scan is submitted on 128 processors.
Then, 8 runs will start simultaneously and finish likely about the same time.
The remaining run is then starting while 112 processors have to idle!
It would in this case be better to either submit on less nodes, or to use more processes per gene run (32 maybe)
although parallel efficiency is not optimal. 
Consider to set  \texttt{--min\_procs} of \texttt{--max\_n\_par\_sims} for the initial efficiency test.


\subsubsection{Autoparallelization and scanscript}

A performance optimization is done before the actual scan. The
result is then used for all following calls of \gene. Even at fixed
values of the \texttt{parallelization} namelist, the parameters \texttt{n\_blocks} and \texttt{perf\_vec} are determined.

The optimal combination of parallelization and \texttt{perf\_vec},
however, changes under the following conditions:
\begin{itemize}
\item Box parameters \texttt{nx0, nky0, nz0, nw0, nv0} are scanned over.
\item \texttt{beta=0}(\texttt{coll=0}) occurs in a \texttt{beta}(\texttt{coll}) scan.
\end{itemize}

In these cases, the scanscript lets the autoparallelization settings
given in the parameters file unchanged.

\subsubsection{Checkpoint reading in scans}

Reading checkpoints is useful in the following cases

\begin{itemize}
\item An unfinished scan is restarted with the \texttt{--continue\_scan}
option. Then the scanscript automatically lets \gene try to read \texttt{s\_checkpoint} files.

\item Checkpoints of finished \gene simulations of a running scan are to
be used as initial conditions for following simulations. This
speeds up computation, in particular for linear eigenvalue runs, but is
not recommended for initial value simulations. Set
\texttt{read\_checkpoint} = T to use this feature.

\item One checkpoint is to be read as an input for all runs of a scan.
Set \texttt{chpt\_in = '<full path>'} in the \texttt{\&scan} namelist and
\texttt{read\_checkpoint} = T in the \texttt{\&in\_out} namelist.
\end{itemize}

\newpage


\section{Output files}
\label{sec:output}

The main output files are the \texttt{nrg}, the
\texttt{field} and the \texttt{mom} files.
The first contains the timetrace information, such as density,
temperatures and the transport fluxes, all spatially averaged.
The \texttt{field} and the \texttt{mom} files contain 3D information
 about fields and moments of the distribution function and become rather large.
For this reason they are only written at some time steps (typically all 100-500).
In addition, some more output files and diagnostics exist which will be explained below.

\subsection{The \texttt{nrg} file}
\label{subsec:nrg-file}

The entries in this file are sequentially written in ASCII for each time step divisible by 
\hyperlink{istep_nrg}{\tt istep\_nrg}. One entry looks like
\begin{center}
\scriptsize \tt
\begin{tabular}{llllllll}
596.505348 & & & & & & &\\
2.4862E+01 & 4.4911E+00 & 2.1926E+01 & 1.2346E+02 & 1.8177E+00 & 1.1874E-02 & \ldots \\ %1.2012E+01 & -3.4358E-02\\
3.5781E+01 & 1.7368E+02 & 2.3356E+01 & 1.2562E+01 & 1.8177E+00 & 1.1874E-02 & \ldots %2.7023E+00 & 2.4421E-02
\end{tabular}
\end{center}
The first line of each data block presents the corresponding simulation time, 
followed by as many lines and in the same order as species are defined in
the {\tt parameters} file. The columns refer to the spatially averaged\footnote{with respect to the full simulation
volume}, normalized fluctuating quantities\\[-5ex]
\begin{center}
\begin{tabular}{*{10}{c}} %{p{2em}}}
$\frac{\avg{\abs{n_1}^2}}{(n_0\rho_{\rf}^*)^2}$ & $\frac{\avg{|u_{1\parallel}|^2}}{(v_T \rho_\rf^*)^2}$ & 
$\frac{{\avg{|T_{1\parallel}|^2}}}{(T_0 \rho_\rf^*)^2}$ & $\frac{{\avg{\abs{T_{1\perp}}^2}}}{(T_0 \rho_\rf^*)^2}$ & 
$\frac{\avg{\Gamma_{\rm es}^x}}{\Gamma_{\rm gb}}$ & $\frac{\avg{\Gamma_{\rm em}^x}}{\Gamma_{\rm gb}}$ &
$\frac{\avg{Q_{\rm es}^x}}{Q_{\rm gb}}$ & $\frac{\avg{Q_{\rm em}^x}}{Q_{\rm gb}}$ &
$\frac{\avg{\Pi_{\rm es}^x}}{\Pi_{\rm gb}}$ & $\frac{\avg{\Pi_{\rm em}^x}}{\Pi_{\rm gb}}$ \\
\end{tabular}
\end{center}
with units being detailed in Sec.~\ref{sec:normalization}. 
The velocity space moments of the fluctuating part of the {\em particle} distribution function $f^{(pc)}$ and the 
particle, heat and parallel momentum fluxes are defined as\\[1ex]
\begin{tabular}{lp{3em}l}
$n_{1} = \int\!\!\D^3v\,\, f_1^{(pc)}$                     & & $\mvec{\Gamma} = \int\!\!\D^3v\,\, f_1^{(pc)}\,\, \mvec{v}_D$ \\
$u_{1\|} = \frac{1}{n_0} \int\!\!\D^3v\,\,v_\| f_1^{(pc)}$ & & $\mvec{Q} = \int\!\!\D^3v\,\, \frac{1}{2}m v^2 f_1^{(pc)}\,\, \mvec{v}_D$ \\
$T_{1\|} = \frac{m}{n_{0}} \int\!\!\D^3v\,\,(v_\|-u_{1\|})^2 f_1^{(pc)} - T_{0} \frac{n_{1}}{n_{0}}$ & & 
$\mvec{\Pi} = \int\!\!\D^3v\,\, m v_{\|} f_1^{(pc)}\,\, \mvec{v}_D$ \\
$T_{1\perp} = \frac{m}{2n_{0}} \int\!\!\D^3v\,\,v_\perp^2 f_1^{(pc)} - T_{0} \frac{n_{1}}{n_0}$ & \\
\end{tabular}\\[1ex]
Here, the drift velocity $\mvec{v}_D$ is approximated by the generalized $\mvec{E}\times\mvec{B}$ velocity.
The fluxes are furthermore split into their electrostatic (es) and electromagnetic (em) fractions and 
projected along the radial direction. The choice of the corresponding contravariant basis vector depends on 
the \hyperlink{norm_flux_projection}{\tt norm\_flux\_projection} switch and can either be 
$\nabla x$ or the unit vector $\nabla x / |\nabla x |$. The advantage of the latter option 
(default in prerelease-1.6 only) is the independence of the radial coordinate while the former 
option is preferable in the context of transport equations where, e.g., the continuity equation yields
\bea
\dot N & = -\int_V\pderiv{n}{t}\D V = \int_S \mvec{\Gamma}\cdot\nabla x/|\nabla x|\D S \nn \\
 & = \int_S \mvec{\Gamma}\cdot\nabla x J \D y\D z \nn
\eea
and hence $\int \dot N \D x = -V \avg{\Gamma^x}$ since $\avg{\ldots } = 1/V \int \ldots J \D x \D y \D z$.
Note that both alternatives can always be restored in the post-processing using the 3D data - the 
statistics might, however, be less accurate.

\subsection{The \texttt{field} file}
\label{subsec:field-file}
In contrast to the \texttt{nrg} file, the \texttt{field} files are binary files.
All entries consist of the real valued step time followed by the three 
dimensional ($k_x$, $k_y$ and $z$ space), complex valued fields
\begin{center}
\begin{tabular}{lll}
$\frac{\phi_1}{(T_\rf/e)\rho_\rf^*}$ & 
$\left\{\frac{A_{1\|}}{B_\rf \rho_\rf \rho_\rf^*} \right.$ &
$\left. \left[ \frac{B_{1\|}}{B_\rf \rho_\rf^*} \right]\right\}$
\end{tabular}
\end{center}
Note that $A_{1\|}$ and $B_{1\|}$ are only appended if actually being computed, 
i.e.~\hyperlink{beta}{\tt beta} has to be larger than zero for both and 
parallel magnetic fluctuations need to be switched on via \hyperlink{bpar}{\tt bpar} 
for the latter.\\
The total number of fields {\tt n\_fields} can also be found in the info namelist
 of the {\sc GENE} output {\tt parameter} file.

\subsection{The \texttt{mom} file}
\label{subsec:mom-file}
\gene produces a \texttt{mom} file for each species which can be distinguished
by the appended species name defined in the corresponding parameter namelist. \\
The \texttt{mom} file structure is very similar to the \texttt{field} file structure.
After the real valued time step information the following three dimensional, 
complex valued {\em velocity space moments}
\begin{center}
\begin{tabular}{llllll}
$\frac{n_1}{n_0 \rho_\rf^*}$ & $\frac{T_{1\parallel}}{T_0 \rho_\rf^*}$ & 
$\frac{T_{1\perp}}{T_0 \rho_\rf^*} $ & $\frac{q_{1\parallel}+1.5 p_0 u_{1\parallel}}{p_0 c_\rf \rho_\rf^*}$ & 
$\frac{q_{1\perp}+p_0 u_{1\parallel}}{p_0 c_\rf \rho_\rf^*}$ & $\frac{u_{1\parallel}}{c_\rf \rho_\rf^*}$  
\end{tabular}
\end{center}
are written to the file. The combinations of the average velocity and the parallel 
and perpendicular components of the parallel heat current density are defined as 
\bea
q_{1\parallel}+1.5 p_0 u_{1\parallel} & = \frac{m}{2} \int\!\!\D^3v\,\,v_\|^3 f^{(pc)} \nn \\
q_{1\perp}+p_0 u_{1\parallel} & = \frac{m}{2} \int\!\!\D^3v\,\,v_\| v_\perp^{2} f^{(pc)} \nn
\eea
where higher orders in perturbed quantities have been neglected.
%\\

%If {\tt trapdiag} is switched on in the {\tt parameter} file, a total of 18 moments will
%be written into the {\tt mom} file. The first six are the passing contributions, the second 
%the trapped and the last six the FLR corrections of moments presented above.\\
%The total number of moments {\tt n\_moms} can also be found in the info namelist
% of the {\sc GENE} output {\tt parameter} file.

\subsection{The \texttt{vsp} file}
\label{subsec:vsp-file}
In contrast to {\tt mom} and {\tt field} files where information about the
three space dimensions are stored, the {\tt vsp} file contains velocity space ($v_\|$, $\mu$)
data dependent on the parallel ($z$) coordinate.
A data block consists of the real valued time step followed by the particle and heat fluxes 
$\Gamma_{\rm es}$, $\Gamma_{\rm em}$, $Q_{\rm es}$, $Q_{\rm em}$ and the distribution function 
$\langle\left|f\right|\rangle^{1/2}_{k_x,k_y}$ averaged over the perpendicular directions.

\subsection{The \texttt{neoclass} file}
\label{subsec:neoclass-file}
The entries of this file are sequentially written in ASCII for each time step divisible by 
\hyperlink{istep_neoclass}{\tt istep\_neoclass} or at the end of a \hyperlink{comp_type}{\rm comp\_type}='NC' run.
One entry looks like
\begin{center}
\scriptsize \tt
\begin{tabular}{llll}
100.0 & & & \\
2.4862E+01 & 4.4911E+00 & 2.1926E+01 & 1.2346E+02 \\
3.5781E+01 & 1.7368E+02 & 2.3356E+01 & 1.2562E+01 
\end{tabular}
\end{center}
The first line of each data block presents the corresponding simulation time, 
followed by as many lines and in the same order as species are defined in
the {\tt parameters} file. The columns refer to the flux-surface averaged, 
normalized neoclassical equilibrium quantities\\[-2ex]
\begin{center}
\begin{tabular}{*{10}{c}} %{p{2em}}}
$\frac{\avg{\Gamma^{\rm neo}}}{\Gamma_{\rm gb}}$ & $\frac{\avg{Q^{\rm neo}}}{Q_{\rm gb}}$ & 
$\frac{{\avg{\Pi^{\rm neo}}}}{\Pi_{\rm gb}}$ & $\frac{{\avg{j_b}}}{(n_{\rf} c_{\rf} B_{\rf} \rho_\rf^*)}$ \\ 
\end{tabular}
\end{center}
with units being detailed in Sec.~\ref{sec:normalization}.
The neoclassical fluxes (and the bootstrap current) are computed as velocity space moments of the
{\em nonfluctuating} part $f_1^{(nc)}$ of the perturbed distribution function,
which has the Fourier component $k_x=k_y=0$. They are defined as\\[1ex]
\begin{tabular}{lp{3em}l}
$\Gamma^{\rm neo} = \int\!\!\D^3v\,\, f_1^{(nc)}\,\, \mvec{v}^r_d$ & & 
$\mvec{\Pi}^{\rm neo} = \int\!\!\D^3v\,\, m v_{\|} f_1^{(nc)}\,\, \mvec{v}^r_d$ \\
$\mvec{Q}^{\rm neo} = \int\!\!\D^3v\,\, \frac{1}{2}m v^2 f_1^{(nc)}\,\, \mvec{v}^r_d$ & & 
$j_{b} = \int\!\!\D^3v\,\,B_0v_\| f_1^{(nc)}$ 
\end{tabular}\\[1ex]
Here, $\mvec{v}_d$ denotes (the radial component of) the combined curvature and $\nabla B$ drifts.
Fluctuating fields, as well as the maxwellian background are not
considered.
The contravariant basis vector used for radial projection depends on 
the \hyperlink{norm_flux_projection}{\tt norm\_flux\_projection} switch, as in the {\tt nrg} file.

\subsection{The \texttt{energy} file}
\label{subsec:energy-file}
This file contains the time evolution of free energy
quantities written in ASCII for each time step divisible by 
\hyperlink{istep_energy}{\tt istep\_energy}. 

We define the free energy operator as the volume average\\
$\mathcal{E}[A] = \sum_j \Re\{\langle n_{0j} T_{0j}\int \mathrm{d}\mu\mathrm{d}v_{\|} h_j/F_{0j}\;
A_j\rangle_{xyz}\}$ summed over species and given in normalized units.
\\ Here, $h_{1j}=F_{1j}+(q_j/T_{0j}\bar{\phi}_1 +\mu
\bar{B}_{1\|})F_{0j}$ is a modified distribution function.
The argument $A_j$ can be the distribution function $F_{1j}$ or
contributions to its time
derivative $\partial_t F_{1j}$ in the gyrokinetic equation.

The energy file has has 14 colums, that enable to diagnose free energy
balance in the simulation. 
The first column gives the simulation time $t$ in units of
$L_\mathrm{ref}/c_\mathrm{ref}$, the second column is the total free
energy ($A_j=F_{1j}$) in units of $E_{\mathrm{ref}}\equiv n_{\mathrm{ref}} T_{\mathrm{ref}}
(\rho^*_{\mathrm{ref}})^2$ and the following columns contain contributions to the free
energy balance ($A_j$ of type $\partial_t F_{1j}$) in units of $E_{\mathrm{ref}}/(L_\mathrm{ref}/c_\mathrm{ref})$.
Details on the normalization are given in
Sec.~\ref{sec:normalization}. In summary, the energy columns are:\\[-5ex]
\begin{center}
\begin{tabular}{lll}
1  & $t$ & simulation time\\
2  & $E_{tot}$ & total free energy\\
3  & $(dE/dt)_{tot}$ & total change of free energy\\
4  & $(dE/dt)_{drive}$ & contribution of the drive term that includes $\nabla_T$ and $\nabla_n$\\
5  & $(dE/dt)_{source}$ &heat and particle source terms (zero in local simulations)\\
6  & $(dE/dt)_{coll}$ &collisional dissipation\\
7  & $(dE/dt)_{Dz}$  & dissipation by z hyperdiffusion\\
8  & $(dE/dt)_{Dv}$  & dissipation by v hyperdiffusion\\
9  & $(dE/dt)_{Dxy}$ & dissipation by x and y hyperdiffusion\\
10 & $(dE/dt)_{nl}$  & contribution of the nonlinear term\\
   &                 & (zero to machine precision due to xy integration)\\
11 & $(dE/dt)_{zv}$  & contribution of the parallel(zv) poisson bracket\\
   &                 & (zero to machine precision unless dissipative z boundary cond.\\
   &                 & or non-standard derivative methods are used in z and/or $v\|$)\\
12 & $(dE/dt)_{rest}$ & curvature and remaining terms \\
   &                 &(zero to machine precision in local simulations)\\
13 & $(dE/dt)_{check}$ & check for conservation:
$((dE/dt)_{zv+nl+rest})/(dE/dt)_{drive}$\\
14 &$(dE/dt)_{tot,2}$ & $(dE/dt)_{tot}$ computed from two consecutive $E_{tot}$ values\\
   &                 &(should be similar to second column,\\
   &                 &depending on \texttt{istep\_energy} and the time integration scheme)\\
\end{tabular}
\end{center}

If the parameter \hyperlink{istep_energy3d}{\tt istep\_energy3d} is set,
energy terms are written into another (binary) energy3d file before they are
volume-averaged in $\{x,y,z\}$ space, when the time index is divisible
by {\tt istep\_energy3d}. Some of the above terms are combined to
keep the file smaller. Each entry begins with the time (a real number)
followed by six (real-valued) energy quantities in ($k_x$, $k_y$, $z$)
space: (i) $E_{tot}$, (ii)
$(dE/dt)_{NC}$ (combining non-conservative terms 4-11, except 9), (iii)
$(dE/dt)_{drive}$, (iv) $(dE/dt)_{coll}$, (v) $(dE/dt)_{diss}$
(combining hyperdiffusion terms 7-9) (vi)$(dE/dt)_{nl}$.They can be analyzed with the
{\sc GENE} diagnostics tool. 

\subsection{The checkpoint files}
\label{subsec:chpt-files}
Depending on the {\tt write\_checkpoint} and {\tt istep\_schpt} settings a {\tt checkpoint} file 
and/or a {\tt s\_checkpoint} (secure checkpoint) file will be written to the {\tt chptdir} 
directory (for further explanation see section \ref{subsubsec:inout}). \\
The files contain the full 5D distribution function and information about the used precision and 
resolutions and will be binary files written with MPI-IO unless chosen otherwise 
(see again section \ref{subsubsec:inout}). Another file with very similar properties but 
containing multiple snapshots of the distribution function is the {\tt g1} file which is 
controlled by {\tt istep\_g1}.

\subsection{The \texttt{omega} file}
\label{subsec:omega-file}

The {\tt omega} file will be written in linear initial value simulations 
if \hyperlink{istep_omega}{\tt istep\_omega} is greater than
zero and consists of a table with three columns
\begin{center}
\begin{tabular}{lll}
$k_y\rho_\rf$ & $\frac{\gamma}{c_\rf/L_\rf}$ & $\frac{\omega}{c_\rf / L_\rf}$
\end{tabular}
\end{center}
for as many $k_y$ values as considered in the simulation (typically one).
If a mode fails to converge up to the precision given by \hyperlink{omega_prec}{\tt omega\_prec}, 
growth rate and real frequency will be set to zero.

\subsection{The \texttt{eigenvalues} file}
\label{subsec:eigenvalues-file}

The {\tt eigenvalues} file will be written at the end of linear eigenvalue simulations 
and consists of a header line specifying the considered $k_y$ value in units of $\rho_\rf$ and 
as many lines as eigenvalues have been found - at least as many as specified by {\tt n\_ev} 
if the code has converged. The output format is
\begin{center}
\begin{tabular}{ll}
$\frac{\gamma}{c_\rf/L_\rf}$ & $\frac{\omega}{c_\rf / L_\rf}$
\end{tabular}
\end{center}


\subsection{The \texttt{autopar} file}
\label{subsec:autopar-file}

The {\tt autopar} file will be written upon start-up when \gene is checking the performance for 
all possible {\tt MPI} setups and the efficiency of different implementations with respect to cache sizes.
A typical entry is
{\scriptsize
\begin{verbatim}
parallelization:    s   v   w   x   y   z
parallelization:    1   1   8   1   1   1  nblocks:    64
   1 1 1 1 1 1 1 1 1 :    4.8712 MB, t =  5.629E-02 sec
   2 1 1 1 1 1 1 1 1 :    7.8877 MB, t =  5.634E-02 sec
   (...)
   1 2 3 1 2 2 2 1 2 :    4.8184 MB, t =  5.332E-02 sec
   best perf_vec      min. WCT     mem_min  mem_max deficit(byte) data segment
   1 2 3 1 2 2 2 1 1  4.4962E-02     3.57     7.89      238032    23.043
parallelization:    1   1   4   1   1   2  nblocks:    64
   (...)
\end{verbatim}}
where the first two lines contain information about the parallelization and the block size (for 
strip mining) being considered in the following section. Afterwards, the wall clock time for each 
individual \hyperlink{perf_vec}{\tt perf\_vec} setting is printed together with a memory estimate 
for those routines. For each parallelization, the best choice is finally redisplayed and 
supplemented by the minimum and maximum memory estimates as well as the actually used work space 
(this, however, might not be available/accurate on every machine).


%\newpage

%\section{Source files}
%\label{sec:source}

\newpage

\section{The Diagnostic Tool}
\label{sec:diag}

For visualizing one's \gene data, be it the fields, moments,
velocity space, or the nrg file data, one can employ the bundled
diagnostic tool. The publicly available version is able to run
under the free IDL Virtual Machine (\url{www.ittvis.com/idlvm}) 
without an IDL licence, and comes with a graphical user interface and an
assortment of standard visualization diagnostics. Please note that
Win32 users will have to create an (empty) file \verb|win32| in
the diagnostics folder in order for the \gene diagnostic tool to
run under their operating system.

\subsection{The graphical user interface}

On a machine from where the data which is to be analyzed can be
accessed, enter the \verb|diagnostics| directory and start IDL VM with \\
{\tt idl -vm=vm\_diag.sav} to open the graphical user interface (GUI). 
The latter contains a display area on the top left, a run 
specification area next to it, the diagnostics table on the bottom, 
and the command buttons on the far right.

In the data path field, enter the (relative or full) path
to the directory in which your run data (\verb|nrg_[run#]| etc.) is saved; the buttons next
to the data path field can be set to default paths for your
convenience (for this, you will be asked to provide your choice
when clicking the buttons for the first time; alternatively you 
can modify the paths in the \verb|internal/guiform| file). 
The next field specifies the directory in which the diagnostic output
(postscript/data/HDF5 files)
is stored. If no path is specified, the default \verb|output|
directory is chosen.

Next, enter the run number in the runs field; if more than one
number is to be specified (i.e., the first run has follow-ups),
either separate the numbers by a comma (e.g.~``50,51,58''), or
supply ranges (e.g.~``50-54''), or use a combination of both. 
If follow-ups are marked with concluding letters in 
alphabetical order (e.g.~``50,50a,50b''), they will automatically
be inserted by using a plus sign (e.g.``50+'') after the base 
run number. In
order to analyze a run which has not been assigned a run number,
simply enter ``act''; this works in combination with the above
methods, e.g.~``50,51+,57-59,60b-d,act''. After entering the run 
numbers, it is recommended that you press the return key. Please 
note that the following file name convention is expected: ``act'' 
expects unchanged file names, e.g. \verb|parameters.dat|,
\verb|field.dat|. Runs are labeled by renaming to
\verb|[base]\_[number]| (e.g.~``field\_50'') or
\verb|[number]\_[base]| (e.g.~``50\_field''). In general, one can
use letters instead of numbers also, if the regular expression
``act'' is avoided.

To specify the time frame which is to be analyzed, use the start
and end fields. For convenience, the first and last buttons look
up the first and last time step, respectively, that is available
from the runs entered in the runs field.

The particle button group is redrawn if one changes the runs
field, for the number and names of the species may change. Select
the species which you wish to analyze here. Next to it is the
normalization drop down menu where different normalizations can be
chosen. L\_perp and R are always available (others depend on the
species), and stand for the standard \gene length scale and the
major plasma radius, respectively. Lastly, there is the output
format button group, where you can choose to create a data file, a
postscript, or both.

The other buttons next to the run specification area have the
following functionalities:
\begin{itemize}
\item show series info: writes some content of the parameters file
to the display area, e.g.~the resolution, gradients, and whether
the run is nonlinear 
\item show nrg data: writes the most relevant data from the nrg 
file to the display area 
\item start nrg diag: creates a postscript file with an nrg data plot 
\item show geometry: plots geometry coefficients used in your
\gene calculation
\end{itemize}

Below, the diagnostics table lists all bundled diagnostic
programs for a specific data loop selected by the tab
- e.g.~to calculate heat and particle flux spectra or 
to draw contour plots of certain variables for mom/field data.
By clicking on the question mark next to the name, one
gets a popup window with a description of the diagnostic and its
parameters. The column to the right of the name shows which
diagnostics are activated and will be run once the save and start
button is clicked -- toggle the on/off status by clicking in this
column. All the other columns contain the diagnostics parameters.
If a diagnostic asks which variables from the moments and field
files to analyze, that information is to be entered here. Some
parameters are on/off switches which can be toggled by clicking
the corresponding field. Please refer to the diagnostic
programmer's guide placed in \verb|diagnostics/doc/| if you plan
to write and include own diagnostic programs.

Again, a range of buttons is provided next to the diagnostics
table:
\begin{itemize}
\item save and start: saves the state of the GUI and starts
reading the run data from the moments and field files and feeds it
to the selected diagnostics; no user action is permitted until the
``finished diagnosing'' dialog window has popped up 
\item load form: loads the most recently saved GUI form, 
e.g.~to undo changes
\item save form: saves the current gui form, e.g.~to save the GUI
state before exiting 
\item clear form: clears all fields in both
the run specification area and the diagnostics table 
\item show
variable list: writes all available moments and fields with their
corresponding variable index to the display area 
\item show custom diags: this button is not available if you 
are using the IDL
Virtual Machine version of the \gene diagnostic. Clicking it will
open an additional window in which one can activate custom
diagnostics. Such diagnostics are to be put into the custom folder
and will be recompiled every time the user clicks the save and
start button. An example is provided with the distribution, and it
is strictly recommended to follow the nomenclature.
\item recent ps: opens the most recently created postscript 
files in ghost view
by spawning a \verb|gv| command in the shell process where the
diagnostic has been started 
\item ps files: opens a file open dialog where the user can 
choose to open any ps files from the
output directory 
\item exits the diagnostic without saving the GUI state
\end{itemize}

Besides displaying data and information, the display can also
be used to abort the diagnosing by clicking and holding the left
mouse button until the next time step is read.

\subsection{The configuration file}

To customize certain functions, a file named \texttt{diag.cfg} can be
edited which is found in the main diagnostics directory. It contains
entries of the form \verb|key = value|. The following keys are
available:
\begin{itemize}
\item \texttt{ps\_viewer [gv]}: one's choice of a viewer for the
postscript output (it is used when one clicks on \verb|recent ps| or
\verb|ps files|)
\item \texttt{vm\_lowcase [no]}: may be set to \verb|yes| to use
lowercase filenames for Virtual Machine output files
\item \texttt{short\_gui [no]}: may be set to \verb|yes| to slightly
reduce the GUI window height---useful in conjunction with certain
screen resolutions
\item \texttt{speedup\_gui [no]}: on certain machines (e.g., HPCFF),
when changing the run number, it takes a long time to refresh the
species namelist; set to \verb|yes| to significantly reduce that delay
\item \texttt{info\_str [2]}: setting for the info string on the
bottom of the output pages (for most diagnostics); 0: no string; 1:
time info only; 2: time info and diag variable settings info
\item \texttt{bg\_color [gray]}: widget background color, requires IDL
restart (it changes a file in the home directory to modify the Motif
settings); see \texttt{internal/global\_vars.pro} for other color
examples
\item \texttt{fg\_color [black]}: widget foreground color, see above
\end{itemize}
If you encounter errors while using the \gene Diagnostics Tool, if you
have modified any of the above entries, please indicate such
modifications when submitting error reports.


\newpage
\section{Tools}
Located in the \texttt{/tools/} directory you can find some 
scripts based on perl or bash which might help you to save time, e.g.
\begin{itemize}
\item \texttt{runassign} to rename the suffix of just generated \gene output 
files (*.dat) to a string provided as second argument
\item \texttt{runrename} to rename the suffix from a string given by the
first argument to a string passed to the script as second argument
\item \texttt{rundelete} to delete a run number range
\item \texttt{runshift} to shift a run number range
\item \texttt{diffnrg} to compare nrg files which for instance 
 have been generated on different machines with the same parameters file.
In contrast to the usual unix diff command \texttt{diffnrg} will ignore
differences smaller than $10^{-15}$.
\item \texttt{gplot} for a simple and quick gnuplot visualization of 
\hyperref[subsec:nrg-file]{nrg} data. Type \texttt{gplot --help} for more 
information.
\end{itemize}

\newpage
\section{Reporting bugs and submitting changes}
Please report bugs or source code modification requests to 
\href{mailto:gene@ipp.mpg.de}{gene@ipp.mpg.de}.
The latter (and possible bug fixes) may be communicated using 
patch files which can be created by calling
\begin{verbatim}
svn diff > patch_release_1_6_rev3855.diff
\end{verbatim}
in your \gene source directory. Please make sure that your 
file name contains the name of the subversion branch (trunk or 
release X.Y) and the revision number (see \verb|svnversion| 
or \verb|svn info|).

\include{gene-faq}

\newpage
\begin{appendix}

\section{Normalizations}\label{sec:normalization}

{\sc Gene}'s implementation of the gyrokinetic set of equations and hence 
all native code output is dimensionless and thus normalized to appropriate quantities.
In this context, the following abbreviations/quantities might appear:\\

\subsection{species dependent}
\begin{tabular}{*{2}{p{3em}l}}
$n_0$& equilibrium density of the {\em corresponding species}\\
$T_0$& equilibrium temperature of the {\em corresponding species}\\
$p_0$& equilibrium 'pressure' of the {\em corresponding species}(!), i.e. $p_0 = n_0 T_0$\\
$m$  & mass of the {\em corresponding species}\\
$v_T$& thermal velocity $v_T = \sqrt{\mathbf{2} T_0/m}$ of the {\em corresponding species}
\end{tabular}

\subsection{reference quantities}
\begin{tabular}{*{2}{p{3em}l}}
$n_\rf$ & density of the {\em reference species} (i.e. species with {\tt dens = 1.0}) \\
$T_\rf$ & reference temperature (temperature of species with {\tt temp = 1.0}) \\
$m_\rf$ & reference mass (mass of species with {\tt mass = 1.0}) \\
$L_\rf$ & reference length (depends on the chosen geometry) \\
$B_\rf$ & reference magnetic field strength (typically on axis; \hyperlink{desc:miller}{miller}: center of flux surface)\\
$q_\rf$ & reference charge $q_\rf = e$\\
$\Omega_{{\rm tor},\rf}$ & reference toroidal angular velocity\\
\end{tabular}\\[1ex]
{\em These reference values are} {\bf not} {\em required for \gene simulations themselves} but can be set 
in the {\tt \&units} namelist (see Sec.~\ref{sec:units_nml}) for post-processing purposes or automatic computation of 
\hyperlink{beta}{\tt beta}, \hyperlink{coll}{\tt coll}, \hyperlink{rhostar}{\tt rhostar} etc.

\subsubsection{derived reference quantities}
\begin{tabular}{*{2}{p{3em}l}}
$p_\rf$ & reference pressure $p_\rf = n_\rf T_\rf$\\
$c_\rf$ & reference velocity $c_\rf = \sqrt{T_\rf/m_\rf}$ (without $\sqrt{2}$!)\\
$\Omega_\rf$ & reference gyrofrequency $\Omega_\rf = q_\rf B_\rf / (m_\rf c)$ \\
$\rho_\rf$ & reference gyroradius $\rho_\rf = c_\rf/\Omega_\rf$ \\
$\rho_\rf^*$ & reference gyroradius-to-machine-size ratio $\rho_\rf^* = \rho_\rf/L_\rf$ \\
$\Gamma_{\rm gb}$ & particle flux GyroBohm units $\Gamma_{\rm gb} = c_\rf n_\rf (\rho_\rf^*)^2$ \\
$Q_{\rm gb}$ & heat flux GyroBohm units $Q_{\rm gb} = c_\rf p_\rf (\rho_\rf^*)^2$ \\
$\Pi_{\rm gb}$ & momentum flux GyroBohm units $\Pi_{\rm gb} = c_\rf^2 m_\rf n_\rf (\rho_\rf^*)^2$\\
\end{tabular}%\\[1ex]

\end{appendix}


\newpage

%\bibliography{literatur}
\begin{thebibliography}{99}
\expandafter\ifx\csname url\endcsname\relax
  \def\url#1{\texttt{#1}}\fi
\expandafter\ifx\csname urlprefix\endcsname\relax\def\urlprefix{URL }\fi
\expandafter\ifx\csname href\endcsname\relax
  \def\href#1#2{#2} \def\path#1{#1}\fi

\bibitem{frieman82} E.~A.~Frieman and L.~Chen, Phys.~Fluids \textbf{25}, 502 (1982)
\bibitem{hahm88} T.~S.~Hahm, Phys.~Fluids \textbf{31}, 2670 (1988)
\bibitem{hahmbriz88} T.~S.~Hahm, W.~W.~Lee, and A.~Brizard, Phys.~Fluids \textbf{31}, 1940 (1988)
\bibitem{brizard89} A.~Brizard, Phys.~Fluids B \textbf{1}, 1381 (1989)
\bibitem{GK07} A.~Brizard and T.~S.~Hahm, Rev.~Mod.~Phys. \textbf{79}, 421 (2007)
\bibitem{gene1} F.~Jenko {\em et al.}, Phys.~Plasmas \textbf{7}, 1904 (2000)
\bibitem{gene2} T.~G\"orler {\em et al.}, J.~Comput.~Phys.~\textbf{230}, 7053 (2011)
%\bibitem[7a]{gene2a} T.~Dannert and F.~Jenko, Phys.~Plasmas \textbf{12}, 072309 (2005)
\bibitem{GENE-webpage} F.~{Jenko} and {The GENE development team}, \href{http://gene.rzg.mpg.de}{The
  {GENE} code}.
\newline\urlprefix\url{http://gene.rzg.mpg.de}
\bibitem{beer} M.~A.~Beer, Ph.D.~Thesis, Princeton (1995)
\bibitem{tracer} P.~Xanthopoulos and F.~Jenko, Phys.~Plasmas \textbf{13}, 092301 (2006)
\bibitem{Arakawa66} A. Arakawa, J.~Comput.~Phys.~\textbf{1}, 119 (1966)
\bibitem{gist} P. Xanthopoulos {\em et al.}, Phys.~Plasmas \textbf{16}, 082303 (2009) 
\bibitem{cyclone} A.~M.~Dimits {\em et al.}, Phys.~Plasmas \textbf{7}, 969 (2000)
\bibitem{petsc1} S.~Balay, K.~Buschelman, V.~Eijkhout, W.~D.~Gropp, D.~Kaushik, 
  M.~G.~Knepley, L.~C.~McInnes, B.~F.~Smith and Hong Zhang, PETSc Users Manual,
  ANL-95/11 - Revision 2.1.5 (2004)
\bibitem{petsc2} S.~Balay, K.~Buschelman, V.~Eijkhout, W.~D.~Gropp, D.~Kaushik,
  M.~G.~Knepley, L.~C.~McInnes, B.~F.~Smith and Hong Zhang, PETSc Web page,
  \urlprefix\url{http://www.mcs.anl.gov/petsc} (2001)
\bibitem{slepc} V.~Hernandez, J.~E.~Roman, and V.~Vidal, ACM Transactions on 
  Mathematical Software, \textbf{31}, 351 (2005)
\bibitem{nrlformulary} J.~D.~Huba, NRL report, \urlprefix\url{http://wwwppd.nrl.navy.mil/nrlformulary} (2011)
\bibitem{HintonHazeltine1976} F.~L.~Hinton and R.~D.~Hazeltine,Rev.~Mod.~Phys. \textbf{48}, 239 (1976)
\bibitem{Hammett06} G.W.~Hammett, W.~Dorland, N.~Loureiro, T.~Tatsuno, APS-DPP Meeting, Philadelphia, Oct. 30 - Nov. 3, 2006. 
\bibitem{diffusion} M.J.~Pueschel, T.~Dannert, and F.~Jenko,
Comput.~Phys.~Commun.~\textbf{181}, 1428 (2010)


\end{thebibliography}
\end{document}

%%% Local Variables: 
%%% mode: latex
%%% TeX-master: t
%%% End: 

